<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Custom Live Plotting for SLiM • slimr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Custom Live Plotting for SLiM">
<meta property="og:description" content="slimr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">slimr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9001</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/Custom_vis.html">Custom Live Plotting for SLiM</a>
    </li>
    <li>
      <a href="../../articles/slimrlang_intro.html">Introduction to slimrlang: Writing SLiM scripts from R</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rdinnager/slimr/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Custom Live Plotting for SLiM</h1>
                        <h4 class="author">Russell Dinnage</h4>
            
            <h4 class="date">22/06/2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rdinnager/slimr/blob/master/vignettes/articles/Custom_vis.Rmd"><code>vignettes/articles/Custom_vis.Rmd</code></a></small>
      <div class="hidden name"><code>Custom_vis.Rmd</code></div>

    </div>

    
    
<style type="text/css" scoped>
PRE.fansi SPAN {padding-top: .25em; padding-bottom: .25em};
</style>
<div id="custom-live-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#custom-live-plots" class="anchor"></a>Custom Live Plots</h2>
<p>This document describes how to make a custom visualization of SLiM simulation results that can be used to monitor the output of simulation while they are running. This takes advantage of the custom callbacks feature of <code><a href="../../reference/slim_run.html">slim_run()</a></code>. As an example we will implements a genome visualization using ‘Hilbert curves’ to visualize the spatial distribution of mutation in a simulation of genome evolution.</p>
</div>
<div id="hilbert-curves" class="section level1">
<h1 class="hasAnchor">
<a href="#hilbert-curves" class="anchor"></a>Hilbert Curves</h1>
<p>Hilbert curves are one example of a “space-filling curve”. Space-filling curves are a way of “folding” a one-dimensional sequence into a two dimensional (or higher) space, in such a way that local spatial structure is maintained (in other words, parts of the sequence that are close together in one dimension also tend to be close together in two dimensions). These are a handy way to visualize a linear sequence (such as a genome) in a more space efficient way, taking advantage of the two dimensions available to us in standard media. An R package for plotting genomic data on Hilbert curves is available (the HilbertCurve package from Bioconductor). We will use a Hilbert curve to visualize the distribution on mutations in a simple genome evolution simulation. We will start be writing our SLiM model using <code>slimr</code> syntax, and using <code><a href="../../reference/slimr_output.html">slimr_output()</a></code> to output the mutations along our simulated genome in our simulated population. This script is based on recipe 6.1.14 from the SLiM manual. It sets up a simulation with three effective “chromosomes” with low recombination within each chromosome. This is the original SLiM script, which we then specify in <code>slimr</code>:</p>
<pre><code>
initialize() {
  initializeMutationRate(1e-7);
  initializeMutationType("m1", 0.5, "f", 0.0);
  initializeMutationType("m2", 0.5, "f", 0.1);
  initializeGenomicElementType("g1", c(m1,m2), c(10000,1));
  initializeGenomicElement(g1, 0, 2999999);
  rates = c(1e-9, 0.5, 1e-9, 0.5, 1e-9);
  ends = c(999999, 1000000, 1999999, 2000000, 2999999);
  initializeRecombinationRate(rates, ends);
}

1 { sim.addSubpop("p1", 500); }

1:10000 late() { sim.outputFULL(); }
</code></pre>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">slimr</span>, <span class="kw">quietly</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">HilbertCurve</span>, <span class="kw">quietly</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">chromosome_sim</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/slim_script.html">slim_script</a></span>(

  <span class="fu"><a href="../../reference/slim_block.html">slim_block</a></span>(<span class="fu">initialize</span>(), {
    <span class="fu"><a href="../../reference/initializeMutationRate.html">initializeMutationRate</a></span>( <span class="fu">slimr_template</span>(<span class="st">"mut_rate"</span>, <span class="fl">1e-7</span>) );
    <span class="fu"><a href="../../reference/initializeMutationType.html">initializeMutationType</a></span>(<span class="st">"m1"</span>, <span class="fl">0.5</span>, <span class="st">"f"</span>, <span class="fl">0.0</span>);
    <span class="fu"><a href="../../reference/initializeMutationType.html">initializeMutationType</a></span>(<span class="st">"m2"</span>, <span class="fl">0.5</span>, <span class="st">"f"</span>, <span class="fl">0.1</span>);
    <span class="fu"><a href="../../reference/initializeGenomicElementType.html">initializeGenomicElementType</a></span>(<span class="st">"g1"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">m1</span>, <span class="no">m2</span>), <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">10000</span>, <span class="fl">1</span>));
    <span class="fu"><a href="../../reference/initializeGenomicElement.html">initializeGenomicElement</a></span>(<span class="no">g1</span>, <span class="fl">0</span>, <span class="fl">262142</span>);
    <span class="no">rates</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1e-9</span>, <span class="fl">0.5</span>, <span class="fl">1e-9</span>, <span class="fl">0.5</span>, <span class="fl">1e-9</span>);
    <span class="no">ends</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">87381</span>, <span class="fl">87382</span>, <span class="fl">174762</span>, <span class="fl">174763</span>, <span class="fl">262143</span>);
    <span class="fu"><a href="../../reference/initializeRecombinationRate.html">initializeRecombinationRate</a></span>(<span class="no">rates</span>, <span class="no">ends</span>);
  }),

  <span class="fu"><a href="../../reference/slim_block.html">slim_block</a></span>(<span class="fl">1</span>, {
    <span class="fu">sim.addSubpop</span>(<span class="st">"p1"</span>, <span class="fu">slimr_template</span>(<span class="st">"N"</span>, <span class="fl">500</span>));
  }),

  <span class="fu"><a href="../../reference/slim_block.html">slim_block</a></span>(<span class="fl">1</span>, <span class="fl">1000</span>, <span class="fu">late</span>(), {
    <span class="fu"><a href="../../reference/slimr_output.html">slimr_output</a></span>( <span class="fu">sim.outputFull</span>() , <span class="st">"fix_mut"</span>, <span class="kw">do_every</span> <span class="kw">=</span> <span class="fl">5</span>);
  })

)

<span class="no">chromosome_sim</span></pre></body></html></div>
<pre class="fansi fansi-output"><code>## &lt;slimr_script[3]&gt;
## <span style="background-color: #00BBBB;font-weight: bold;">block_init:</span><span>initialize() </span><span style="color: #BBBB00;">{</span><span>
##     </span><span style="color: #00BBBB;">initializeMutationRate</span><span style="color: #0000BB;">(</span><span style="color: #00BB00;">..mut_rate..</span><span style="color: #0000BB;">)</span><span>;
##     </span><span style="color: #00BBBB;">initializeMutationType</span><span style="color: #0000BB;">(</span><span style="color: #BBBB00;">"m1"</span><span>, </span><span style="color: #0000BB;">0.5</span><span>, </span><span style="color: #BBBB00;">"f"</span><span>, </span><span style="color: #0000BB;">0)</span><span>;
##     </span><span style="color: #00BBBB;">initializeMutationType</span><span style="color: #0000BB;">(</span><span style="color: #BBBB00;">"m2"</span><span>, </span><span style="color: #0000BB;">0.5</span><span>, </span><span style="color: #BBBB00;">"f"</span><span>, </span><span style="color: #0000BB;">0.1)</span><span>;
##     </span><span style="color: #00BBBB;">initializeGenomicElementType</span><span style="color: #0000BB;">(</span><span style="color: #BBBB00;">"g1"</span><span>, </span><span style="color: #00BBBB;">c(</span><span>m1, m2</span><span style="color: #00BBBB;">)</span><span>, </span><span style="color: #00BBBB;">c(</span><span style="color: #0000BB;">10000</span><span>, </span><span style="color: #0000BB;">1</span><span style="color: #00BBBB;">)</span><span style="color: #0000BB;">)</span><span>;
##     </span><span style="color: #00BBBB;">initializeGenomicElement</span><span style="color: #0000BB;">(</span><span>g1, </span><span style="color: #0000BB;">0</span><span>, </span><span style="color: #0000BB;">262142)</span><span>;
##     rates </span><span style="color: #00BB00;">=</span><span> </span><span style="color: #00BBBB;">c</span><span style="color: #0000BB;">(1e-09</span><span>, </span><span style="color: #0000BB;">0.5</span><span>, </span><span style="color: #0000BB;">1e-09</span><span>, </span><span style="color: #0000BB;">0.5</span><span>, </span><span style="color: #0000BB;">1e-09)</span><span>;
##     ends </span><span style="color: #00BB00;">=</span><span> </span><span style="color: #00BBBB;">c</span><span style="color: #0000BB;">(87381</span><span>, </span><span style="color: #0000BB;">87382</span><span>, </span><span style="color: #0000BB;">174762</span><span>, </span><span style="color: #0000BB;">174763</span><span>, </span><span style="color: #0000BB;">262143)</span><span>;
##     </span><span style="color: #00BBBB;">initializeRecombinationRate</span><span style="color: #0000BB;">(</span><span>rates, ends</span><span style="color: #0000BB;">)</span><span>;
## </span><span style="color: #BBBB00;">}</span><span>
## 
## </span><span style="background-color: #00BBBB;font-weight: bold;">block_2:</span><span>1 early() </span><span style="color: #BBBB00;">{</span><span>
##     </span><span style="color: #00BBBB;">sim.addSubpop</span><span style="color: #0000BB;">(</span><span style="color: #BBBB00;">"p1"</span><span>, </span><span style="color: #00BB00;">..N..</span><span style="color: #0000BB;">)</span><span>;
## </span><span style="color: #BBBB00;">}</span><span>
## 
## </span><span style="background-color: #00BBBB;font-weight: bold;">block_3:</span><span>1:1000 late() </span><span style="color: #BBBB00;">{</span><span>
##     </span><span style="color: #BB0000;">if</span><span> </span><span style="color: #0000BB;">(</span><span>sim.generation </span><span style="color: #00BB00;">%%</span><span> </span><span style="color: #0000BB;">5</span><span> </span><span style="color: #00BB00;">==</span><span> </span><span style="color: #0000BB;">0)</span><span> </span><span style="color: #0000BB;">{</span><span>
##       </span><span style="color: #00BBBB;">cat(</span><span style="color: #BBBB00;">"&lt;slimr_out:start&gt;"</span><span> </span><span style="color: #00BB00;">+</span><span> </span><span style="color: #00BBBB;">paste</span><span style="color: #BBBB00;">(</span><span>sim.generation</span><span style="color: #BBBB00;">)</span><span> </span><span style="color: #00BB00;">+</span><span> </span><span style="color: #BBBB00;">",'"</span><span> </span><span style="color: #00BB00;">+</span><span> </span><span style="color: #BBBB00;">"fix_mut"</span><span> </span><span style="color: #00BB00;">+</span><span> </span><span style="color: #BBBB00;">"','"</span><span> </span><span style="color: #00BB00;">+</span><span>
##         </span><span style="color: #BBBB00;">"sim.outputFull()"</span><span> </span><span style="color: #00BB00;">+</span><span> </span><span style="color: #BBBB00;">"','"</span><span style="color: #00BBBB;">)</span><span>;
##       </span><span style="color: #00BBBB;">sim.outputFull()</span><span>;
##       </span><span style="color: #00BBBB;">cat(</span><span style="color: #BBBB00;">"'&lt;slimr_out:end&gt;"</span><span style="color: #00BBBB;">)</span><span>;
##     </span><span style="color: #0000BB;">}</span><span>
## </span><span style="color: #BBBB00;">}</span><span>
## This slimr_script has templating in block(s) </span><span style="background-color: #00BBBB;font-weight: bold;">block_init</span><span> and
## </span><span style="background-color: #00BBBB;font-weight: bold;">block_2</span><span> for variables </span><span style="color: #00BB00;">mut_rate</span><span> and </span><span style="color: #00BB00;">N</span><span>.
</span></code></pre>
<p>I’ve added a few <code>slimr_template()</code> calls to make it easier to quickly change parameters in the script for exploration.</p>
<p>In order to make a custom visualization for this script we can provide a custom callback function. This is a function which takes as its first argument a dataset named <code>data</code>, which is dynamically constructed while SLiM is running based on the <code><a href="../../reference/slimr_output.html">slimr_output()</a></code> call in the script. To get an example dataset we can use to test our callback function, we can just run the script with <code>capture_output = TRUE</code>, which will capture the data and output it in the final results. This will tell us what the data will look like during the SLiM run. Since we just need it for testing purposes, we can just run the simulation for a small number of generations and on a small number of individuals (just enough to accumulate a few mutations to test the visualization).</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">chromosome_sim_test</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/slimr_script_render.html">slimr_script_render</a></span>(<span class="no">chromosome_sim</span>,
                                           <span class="kw">template</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">N</span> <span class="kw">=</span> <span class="fl">100</span>))</pre></body></html></div>
<pre><code>## Warning in replace_double_dots(slimr_script, .x, slimr_template_attr = slimr_template_attr, : Warning: A templated variable was not specified in the template and has been replaced by its default value.</code></pre>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">output_test</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/slim_run.html">slim_run</a></span>(<span class="no">chromosome_sim_test</span>)</pre></body></html></div>
<pre><code>## 
## 
## Simulation finished with exit status: 0</code></pre>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">output_test</span>$<span class="no">output_data</span></pre></body></html></div>
<pre class="fansi fansi-output"><code>## <span style="color: #555555;"># A tibble: 198 x 4</span><span>
##    generation name    expression    data                                        
##         </span><span style="color: #555555;font-style: italic;">&lt;int&gt;</span><span> </span><span style="color: #555555;font-style: italic;">&lt;chr&gt;</span><span>   </span><span style="color: #555555;font-style: italic;">&lt;chr&gt;</span><span>         </span><span style="color: #555555;font-style: italic;">&lt;chr&gt;</span><span>                                       
## </span><span style="color: #555555;"> 1</span><span>          5 fix_mut sim.outputFu~ </span><span style="color: #555555;">"</span><span>#OUT: 5 A\nVersion: 3\nPopulations:\np1 10~
## </span><span style="color: #555555;"> 2</span><span>         10 fix_mut sim.outputFu~ </span><span style="color: #555555;">"</span><span>#OUT: 10 A\nVersion: 3\nPopulations:\np1 1~
## </span><span style="color: #555555;"> 3</span><span>         15 fix_mut sim.outputFu~ </span><span style="color: #555555;">"</span><span>#OUT: 15 A\nVersion: 3\nPopulations:\np1 1~
## </span><span style="color: #555555;"> 4</span><span>         20 fix_mut sim.outputFu~ </span><span style="color: #555555;">"</span><span>#OUT: 20 A\nVersion: 3\nPopulations:\np1 1~
## </span><span style="color: #555555;"> 5</span><span>         25 fix_mut sim.outputFu~ </span><span style="color: #555555;">"</span><span>#OUT: 25 A\nVersion: 3\nPopulations:\np1 1~
## </span><span style="color: #555555;"> 6</span><span>         30 fix_mut sim.outputFu~ </span><span style="color: #555555;">"</span><span>#OUT: 30 A\nVersion: 3\nPopulations:\np1 1~
## </span><span style="color: #555555;"> 7</span><span>         35 fix_mut sim.outputFu~ </span><span style="color: #555555;">"</span><span>#OUT: 35 A\nVersion: 3\nPopulations:\np1 1~
## </span><span style="color: #555555;"> 8</span><span>         40 fix_mut sim.outputFu~ </span><span style="color: #555555;">"</span><span>#OUT: 40 A\nVersion: 3\nPopulations:\np1 1~
## </span><span style="color: #555555;"> 9</span><span>         45 fix_mut sim.outputFu~ </span><span style="color: #555555;">"</span><span>#OUT: 45 A\nVersion: 3\nPopulations:\np1 1~
## </span><span style="color: #555555;">10</span><span>         50 fix_mut sim.outputFu~ </span><span style="color: #555555;">"</span><span>#OUT: 50 A\nVersion: 3\nPopulations:\np1 1~
## </span><span style="color: #555555;"># ... with 188 more rows</span><span>
</span></code></pre>
<p>So now we can write a function that takes the outputted <code>tibble</code> and make the desired plot. The first thing we need to do is extract the relevant information from the full output. In this case we only need the information on mutations: their position, and their prevalance in the population. We can use the handy <code>slimr</code> function <code><a href="../../reference/slim_outputFull_extract.html">slim_outputFull_extract(type = "mutation")</a></code> for this. This function extracts various pieces of the SLiM <code><a href="../../reference/outputFull.html">outputFull()</a></code> function, which contains all information about a population in a SLiM simulation. In this case, we only need the information for the latest generation. Let’s test it:</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">mut_dat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/slim_outputFull_extract.html">slim_outputFull_extract</a></span>(<span class="no">output_test</span>$<span class="no">output_data</span>[<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">output_test</span>$<span class="no">output_data</span>), ],
                                   <span class="st">"mutations"</span>)
<span class="no">mut_dat</span></pre></body></html></div>
<pre class="fansi fansi-output"><code>## <span style="color: #555555;"># A tibble: 68 x 11</span><span>
##    generation    id unique_id mut_type chrome_pos selection dominance subpop
##         </span><span style="color: #555555;font-style: italic;">&lt;int&gt;</span><span> </span><span style="color: #555555;font-style: italic;">&lt;int&gt;</span><span>     </span><span style="color: #555555;font-style: italic;">&lt;int&gt;</span><span> </span><span style="color: #555555;font-style: italic;">&lt;chr&gt;</span><span>         </span><span style="color: #555555;font-style: italic;">&lt;int&gt;</span><span>     </span><span style="color: #555555;font-style: italic;">&lt;dbl&gt;</span><span>     </span><span style="color: #555555;font-style: italic;">&lt;dbl&gt;</span><span> </span><span style="color: #555555;font-style: italic;">&lt;chr&gt;</span><span> 
## </span><span style="color: #555555;"> 1</span><span>        990    29      </span><span style="text-decoration: underline;">1</span><span>169 m1           </span><span style="text-decoration: underline;">144</span><span>771         0       0.5 p1    
## </span><span style="color: #555555;"> 2</span><span>        990     1      </span><span style="text-decoration: underline;">1</span><span>201 m1           </span><span style="text-decoration: underline;">131</span><span>667         0       0.5 p1    
## </span><span style="color: #555555;"> 3</span><span>        990     0      </span><span style="text-decoration: underline;">1</span><span>691 m1           </span><span style="text-decoration: underline;">120</span><span>338         0       0.5 p1    
## </span><span style="color: #555555;"> 4</span><span>        990    24      </span><span style="text-decoration: underline;">2</span><span>472 m1            </span><span style="text-decoration: underline;">94</span><span>405         0       0.5 p1    
## </span><span style="color: #555555;"> 5</span><span>        990    25      </span><span style="text-decoration: underline;">2</span><span>735 m1            </span><span style="text-decoration: underline;">96</span><span>431         0       0.5 p1    
## </span><span style="color: #555555;"> 6</span><span>        990    26      </span><span style="text-decoration: underline;">3</span><span>081 m1           </span><span style="text-decoration: underline;">107</span><span>929         0       0.5 p1    
## </span><span style="color: #555555;"> 7</span><span>        990    28      </span><span style="text-decoration: underline;">3</span><span>831 m1           </span><span style="text-decoration: underline;">136</span><span>347         0       0.5 p1    
## </span><span style="color: #555555;"> 8</span><span>        990    27      </span><span style="text-decoration: underline;">3</span><span>872 m1           </span><span style="text-decoration: underline;">128</span><span>901         0       0.5 p1    
## </span><span style="color: #555555;"> 9</span><span>        990     3      </span><span style="text-decoration: underline;">4</span><span>334 m1           </span><span style="text-decoration: underline;">228</span><span>378         0       0.5 p1    
## </span><span style="color: #555555;">10</span><span>        990     4      </span><span style="text-decoration: underline;">4</span><span>373 m1             </span><span style="text-decoration: underline;">7</span><span>937         0       0.5 p1    
## </span><span style="color: #555555;"># ... with 58 more rows, and 3 more variables: first_gen </span><span style="color: #555555;font-style: italic;">&lt;int&gt;</span><span style="color: #555555;">,</span><span>
## </span><span style="color: #555555;">#   prevalence </span><span style="color: #555555;font-style: italic;">&lt;int&gt;</span><span style="color: #555555;">, nucleotide </span><span style="color: #555555;font-style: italic;">&lt;chr&gt;</span><span>
</span></code></pre>
<p>Okay, so now we are ready to make our callback function. First we will create a little function to plot the Hilbert Curve. This uses a trick from the <code>grDevices</code> package to hold the display until each plot is finished drawing. This gives the animation produced while running the simulation a smoother appearance, but is not strictly necessary.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="co">## specify chromosome ranges for plotting</span>
<span class="no">chrs</span> <span class="kw">&lt;-</span> <span class="kw pkg">IRanges</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html">IRanges</a></span>(<span class="kw">start</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">87382</span>, <span class="fl">174763</span>),
                         <span class="kw">end</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">87382</span>, <span class="fl">174763</span>, <span class="fl">262143</span>))

<span class="no">bg</span> <span class="kw">&lt;-</span> <span class="kw pkg">IRanges</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html">IRanges</a></span>(<span class="kw">start</span> <span class="kw">=</span> <span class="fl">0</span>, <span class="kw">end</span> <span class="kw">=</span> <span class="fl">262143</span>)

<span class="no">plot_HC</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">positions</span>, <span class="no">prevalence</span>, <span class="no">chrs</span>, <span class="no">bg</span>, <span class="no">max_prev</span>) {
  <span class="no">prevalence</span> <span class="kw">&lt;-</span> <span class="no">prevalence</span> / <span class="no">max_prev</span>
  <span class="no">cols</span> <span class="kw">&lt;-</span> <span class="kw pkg">colourvalues</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/colourvalues/man/colour_values.html">color_values</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="no">prevalence</span>, <span class="fl">1</span>))[<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(-<span class="fl">1</span>, -<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">prevalence</span>))]
  <span class="kw pkg">grDevices</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.flush.html">dev.hold</a></span>()
  <span class="no">hc</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HilbertCurve/man/HilbertCurve.html">HilbertCurve</a></span>(<span class="fl">0</span>, <span class="fl">262143</span>, <span class="kw">level</span> <span class="kw">=</span> <span class="fl">4</span>, <span class="kw">reference</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">arrow</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
  <span class="fu"><a href="https://rdrr.io/pkg/HilbertCurve/man/hc_polygon-dispatch.html">hc_polygon</a></span>(<span class="no">hc</span>, <span class="no">chrs</span>)
  <span class="fu"><a href="https://rdrr.io/pkg/HilbertCurve/man/hc_points-dispatch.html">hc_points</a></span>(<span class="no">hc</span>, <span class="no">bg</span>, <span class="kw">np</span> <span class="kw">=</span> <span class="fl">4</span>, <span class="kw">gp</span> <span class="kw">=</span> <span class="fu">gpar</span>(<span class="kw">fill</span> <span class="kw">=</span> <span class="kw pkg">colourvalues</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/colourvalues/man/colour_values.html">color_values</a></span>(<span class="fl">0</span>)))
  <span class="fu"><a href="https://rdrr.io/pkg/HilbertCurve/man/hc_points-dispatch.html">hc_points</a></span>(<span class="no">hc</span>, <span class="kw">x1</span> <span class="kw">=</span> <span class="no">positions</span>, <span class="kw">gp</span> <span class="kw">=</span> <span class="fu">gpar</span>(<span class="kw">fill</span> <span class="kw">=</span> <span class="no">cols</span>), <span class="kw">np</span> <span class="kw">=</span> <span class="fl">4</span>, <span class="kw">mean_mode</span> <span class="kw">=</span> <span class="st">"weighted"</span>)
  <span class="fu"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="kw pkg">grDevices</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.flush.html">dev.flush</a></span>())
}

<span class="co">#grDevices::dev.new(noRStudioGD = TRUE)</span>
<span class="fu">plot_HC</span>(<span class="no">mut_dat</span>$<span class="no">chrome_pos</span>, <span class="no">mut_dat</span>$<span class="no">prevalence</span>, <span class="no">chrs</span>, <span class="no">bg</span>, <span class="kw">max_prev</span> <span class="kw">=</span> <span class="fl">200</span>)</pre></body></html></div>
<p><img src="Custom_vis_files/figure-html/make_callback-1.png" width="700"></p>
<p>Now the final callback function:</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">make_hilbert</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">data</span>, <span class="no">max_prev</span>) {
  <span class="no">mutation_data</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/slim_outputFull_extract.html">slim_outputFull_extract</a></span>(<span class="no">data</span>[<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">data</span>), ], <span class="st">"mutations"</span>)
  <span class="fu">plot_HC</span>(<span class="no">mutation_data</span>$<span class="no">chrome_pos</span>, <span class="no">mutation_data</span>$<span class="no">prevalence</span>, <span class="no">chrs</span>, <span class="no">bg</span>, <span class="no">max_prev</span>)
  <span class="fu"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="kw">NULL</span>)
}</pre></body></html></div>
<p>Let’s try it out!</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co">## change mutation rate and increase pop size</span>
<span class="no">chromosome_sim_1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/slimr_script_render.html">slimr_script_render</a></span>(<span class="no">chromosome_sim</span>,
                                           <span class="kw">template</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">N</span> <span class="kw">=</span> <span class="fl">500</span>,
                                                           <span class="kw">mut_rate</span> <span class="kw">=</span> <span class="fl">1e-6</span>))

<span class="co">## increase generations</span>
<span class="fu">modify</span>(<span class="no">chromosome_sim_1</span>, <span class="st">"end_gen"</span>, <span class="fl">3</span>) <span class="kw">&lt;-</span> <span class="fl">10000</span>

<span class="co">## open a non-rstudio graphics device (faster and support holding)</span>
<span class="kw pkg">grDevices</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.new</a></span>(<span class="kw">noRStudioGD</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="no">sim_results</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../../reference/slim_run.html">slim_run</a></span>(<span class="no">chromosome_sim_1</span>,
                        <span class="kw">callbacks</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">make_hilbert</span>),
                        <span class="kw">cb_args</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">max_prev</span> <span class="kw">=</span> <span class="fl">1000</span>))</pre></body></html></div>
<p><img src="figures/HC_callback_vis_demo.gif"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Russell Dinnage, Stephen Sarre, Bernd Gruber, Richard Duncan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
