<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Custom Live Plotting for SLiM • slimr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Custom Live Plotting for SLiM">
<meta property="og:description" content="slimr">
<meta property="og:image" content="https://rdinnager.github.io/slimr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">slimr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Custom_vis.html">Custom Live Plotting for SLiM</a>
    </li>
    <li>
      <a href="../articles/Main_manuscript_example_v2.html">Using slimr to investigate the population genetics of the Sandy Inland Mouse in the periodic rainfall environment of the Simpson Desert</a>
    </li>
    <li>
      <a href="../articles/multispecies.html">Advanced Example: Multispecies Model</a>
    </li>
    <li>
      <a href="../articles/simple_example_using_migration_and_fst.html">simple_example_using_migration_and_fst</a>
    </li>
    <li>
      <a href="../articles/installing_SLiM_for_slimr.html">How to Install SLiM for Use with slimr</a>
    </li>
    <li>
      <a href="../articles/simple_nucleotide_example.html">Simple Example of a Simulation using Nucleotides</a>
    </li>
    <li>
      <a href="../articles/slimrlang_intro.html">Introduction to slimr: Writing SLiM scripts from R</a>
    </li>
    <li>
      <a href="../articles/useful_tables.html">Useful slimr Tables</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rdinnager/slimr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Custom Live Plotting for SLiM</h1>
                        <h4 data-toc-skip class="author">Russell
Dinnage</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/rdinnager/slimr/blob/HEAD/vignettes/articles/Custom_vis.Rmd" class="external-link"><code>vignettes/articles/Custom_vis.Rmd</code></a></small>
      <div class="hidden name"><code>Custom_vis.Rmd</code></div>

    </div>

    
    
<style type="text/css" scoped>
PRE.fansi SPAN {padding-top: .25em; padding-bottom: .25em};
</style>
<div class="section level3">
<h3 id="custom-live-plots">Custom Live Plots<a class="anchor" aria-label="anchor" href="#custom-live-plots"></a>
</h3>
<p>This document describes how to make a custom visualization of SLiM
simulation results that can be used to monitor the output of simulation
while they are running. This takes advantage of the custom callbacks
feature of <code><a href="../reference/slim_run.html">slim_run()</a></code>. As an example we will implements a
genome visualization using ‘Hilbert curves’ to visualize the spatial
distribution of mutation in a simulation of genome evolution.</p>
</div>
<div class="section level2">
<h2 id="hilbert-curves">Hilbert Curves<a class="anchor" aria-label="anchor" href="#hilbert-curves"></a>
</h2>
<p>Hilbert curves are one example of a “space-filling curve”.
Space-filling curves are a way of “folding” a one-dimensional sequence
into a two dimensional (or higher) space, in such a way that local
spatial structure is maintained (in other words, parts of the sequence
that are close together in one dimension also tend to be close together
in two dimensions). These are a handy way to visualize a linear sequence
(such as a genome) in a more space efficient way, taking advantage of
the two dimensions available to us in standard media. An R package for
plotting genomic data on Hilbert curves is available (the HilbertCurve
package from Bioconductor). We will use a Hilbert curve to visualize the
distribution on mutations in a simple genome evolution simulation. We
will start be writing our SLiM model using <code>slimr</code> syntax,
and using <code><a href="../reference/r_output.html">slimr_output()</a></code> to output the mutations along our
simulated genome in our simulated population. This script is based on
recipe 6.1.14 from the SLiM manual. It sets up a simulation with three
effective “chromosomes” with low recombination within each chromosome.
This is the original SLiM script, which we then specify in
<code>slimr</code>:</p>
<pre><code>
initialize() {
  initializeMutationRate(1e-7);
  initializeMutationType("m1", 0.5, "f", 0.0);
  initializeMutationType("m2", 0.5, "f", 0.1);
  initializeGenomicElementType("g1", c(m1,m2), c(10000,1));
  initializeGenomicElement(g1, 0, 2999999);
  rates = c(1e-9, 0.5, 1e-9, 0.5, 1e-9);
  ends = c(999999, 1000000, 1999999, 2000000, 2999999);
  initializeRecombinationRate(rates, ends);
}

1 { sim.addSubpop("p1", 500); }

1:10000 late() { sim.outputFULL(); }
</code></pre>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rdinnager.github.io/slimr">slimr</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/HilbertCurve" class="external-link">HilbertCurve</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chromosome_sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim_script.html">slim_script</a></span><span class="op">(</span></span>
<span>  </span>
<span>  <span class="fu"><a href="../reference/slim_block.html">slim_block</a></span><span class="op">(</span><span class="fu"><a href="../reference/callback_initialize.html">initialize</a></span><span class="op">(</span><span class="op">)</span>, <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/initializeMutationRate.html">initializeMutationRate</a></span><span class="op">(</span> <span class="fu"><a href="../reference/r_template.html">slimr_template</a></span><span class="op">(</span><span class="st">"mut_rate"</span>, <span class="fl">1e-7</span><span class="op">)</span> <span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeMutationType.html">initializeMutationType</a></span><span class="op">(</span><span class="st">"m1"</span>, <span class="fl">0.5</span>, <span class="st">"f"</span>, <span class="fl">0.0</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeMutationType.html">initializeMutationType</a></span><span class="op">(</span><span class="st">"m2"</span>, <span class="fl">0.5</span>, <span class="st">"f"</span>, <span class="fl">0.1</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeGenomicElementType.html">initializeGenomicElementType</a></span><span class="op">(</span><span class="st">"g1"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">m1</span>, <span class="va">m2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10000</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeGenomicElement.html">initializeGenomicElement</a></span><span class="op">(</span><span class="va">g1</span>, <span class="fl">0</span>, <span class="fl">262142</span><span class="op">)</span>;</span>
<span>    <span class="va">rates</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1e-9</span>, <span class="fl">0.5</span>, <span class="fl">1e-9</span>, <span class="fl">0.5</span>, <span class="fl">1e-9</span><span class="op">)</span>;</span>
<span>    <span class="va">ends</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">87381</span>, <span class="fl">87382</span>, <span class="fl">174762</span>, <span class="fl">174763</span>, <span class="fl">262143</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeRecombinationRate.html">initializeRecombinationRate</a></span><span class="op">(</span><span class="va">rates</span>, <span class="va">ends</span><span class="op">)</span>;</span>
<span>  <span class="op">}</span><span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="fu"><a href="../reference/slim_block.html">slim_block</a></span><span class="op">(</span><span class="fl">1</span>, <span class="op">{</span></span>
<span>    <span class="fu">sim.addSubpop</span><span class="op">(</span><span class="st">"p1"</span>, <span class="fu"><a href="../reference/r_template.html">slimr_template</a></span><span class="op">(</span><span class="st">"N"</span>, <span class="fl">500</span><span class="op">)</span><span class="op">)</span>;</span>
<span>  <span class="op">}</span><span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="fu"><a href="../reference/slim_block.html">slim_block</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1000</span>, <span class="fu"><a href="../reference/late.html">late</a></span><span class="op">(</span><span class="op">)</span>, <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/r_output.html">slimr_output</a></span><span class="op">(</span> <span class="fu">sim.outputFull</span><span class="op">(</span><span class="op">)</span> , <span class="st">"fix_mut"</span>, do_every <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>;</span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">chromosome_sim</span></span></code></pre></div>
<pre class="fansi fansi-output"><code><span><span class="co">## &lt;slimr_script[3]&gt;</span></span>
<span><span class="co">## block_init:initialize() {</span></span>
<span><span class="co">##     initializeMutationRate(..mut_rate..);</span></span>
<span><span class="co">##     initializeMutationType("m1", 0.5, "f", 0);</span></span>
<span><span class="co">##     initializeMutationType("m2", 0.5, "f", 0.1);</span></span>
<span><span class="co">##     initializeGenomicElementType("g1", c(m1, m2), c(10000, 1));</span></span>
<span><span class="co">##     initializeGenomicElement(g1, 0, 262142);</span></span>
<span><span class="co">##     rates = c(1e-09, 0.5, 1e-09, 0.5, 1e-09);</span></span>
<span><span class="co">##     ends = c(87381, 87382, 174762, 174763, 262143);</span></span>
<span><span class="co">##     initializeRecombinationRate(rates, ends);</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## block_2:1 early() {</span></span>
<span><span class="co">##     sim.addSubpop("p1", ..N..);</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## block_3:1:1000 late() {</span></span>
<span><span class="co">##     {sim.outputFull() -&gt; fix_mut}</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## This slimr_script has templating in block(s) block_init and</span></span>
<span><span class="co">## block_2 for variables mut_rate and N.</span></span></code></pre>
<p>I’ve added a few <code><a href="../reference/r_template.html">slimr_template()</a></code> calls to make it
easier to quickly change parameters in the script for exploration.</p>
<p>In order to make a custom visualization for this script we can
provide a custom callback function. This is a function which takes as
its first argument a dataset named <code>data</code>, which is
dynamically constructed while SLiM is running based on the
<code><a href="../reference/r_output.html">slimr_output()</a></code> call in the script. To get an example
dataset we can use to test our callback function, we can just run the
script with <code>capture_output = TRUE</code>, which will capture the
data and output it in the final results. This will tell us what the data
will look like during the SLiM run. Since we just need it for testing
purposes, we can just run the simulation for a small number of
generations and on a small number of individuals (just enough to
accumulate a few mutations to test the visualization).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chromosome_sim_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim_script_render.html">slim_script_render</a></span><span class="op">(</span><span class="va">chromosome_sim</span>,</span>
<span>                                           template <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Warning: A templated variable was not specified in the template and has been replaced by its default value.</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim_run.html">slim_run</a></span><span class="op">(</span><span class="va">chromosome_sim_test</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Simulation finished with exit status: 0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Success!</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_test</span><span class="op">$</span><span class="va">output_data</span></span></code></pre></div>
<pre class="fansi fansi-output"><code><span><span class="co">## # A tibble: 200 × 5</span></span>
<span><span class="co">##    generation name    expression       type        data                         </span></span>
<span><span class="co">##         &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;            &lt;chr&gt;       &lt;chr&gt;                        </span></span>
<span><span class="co">##  1          5 fix_mut sim.outputFull() slim_output "#OUT: 5 5 A\nVersion: 3\nPo…</span></span>
<span><span class="co">##  2         10 fix_mut sim.outputFull() slim_output "#OUT: 10 10 A\nVersion: 3\n…</span></span>
<span><span class="co">##  3         15 fix_mut sim.outputFull() slim_output "#OUT: 15 15 A\nVersion: 3\n…</span></span>
<span><span class="co">##  4         20 fix_mut sim.outputFull() slim_output "#OUT: 20 20 A\nVersion: 3\n…</span></span>
<span><span class="co">##  5         25 fix_mut sim.outputFull() slim_output "#OUT: 25 25 A\nVersion: 3\n…</span></span>
<span><span class="co">##  6         30 fix_mut sim.outputFull() slim_output "#OUT: 30 30 A\nVersion: 3\n…</span></span>
<span><span class="co">##  7         35 fix_mut sim.outputFull() slim_output "#OUT: 35 35 A\nVersion: 3\n…</span></span>
<span><span class="co">##  8         40 fix_mut sim.outputFull() slim_output "#OUT: 40 40 A\nVersion: 3\n…</span></span>
<span><span class="co">##  9         45 fix_mut sim.outputFull() slim_output "#OUT: 45 45 A\nVersion: 3\n…</span></span>
<span><span class="co">## 10         50 fix_mut sim.outputFull() slim_output "#OUT: 50 50 A\nVersion: 3\n…</span></span>
<span><span class="co">## # ℹ 190 more rows</span></span></code></pre>
<p>So now we can write a function that takes the outputted
<code>tibble</code> and make the desired plot. The first thing we need
to do is extract the relevant information from the full output. In this
case we only need the information on mutations: their position, and
their prevalance in the population. We can use the handy
<code>slimr</code> function
<code>slim_extract_full(type = "mutation")</code> for this. This
function extracts various pieces of the SLiM <code><a href="../reference/outputFull.html">outputFull()</a></code>
function, which contains all information about a population in a SLiM
simulation. In this case, we only need the information for the latest
generation. Let’s test it:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mut_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim_extract_full.html">slim_extract_full</a></span><span class="op">(</span><span class="va">output_test</span><span class="op">$</span><span class="va">output_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">output_test</span><span class="op">$</span><span class="va">output_data</span><span class="op">)</span>, <span class="op">]</span>,</span>
<span>                                   <span class="st">"mutations"</span><span class="op">)</span></span>
<span><span class="va">mut_dat</span></span></code></pre></div>
<pre class="fansi fansi-output"><code><span><span class="co">## # A tibble: 72 × 11</span></span>
<span><span class="co">##    generation mut_id unique_mut_id mut_type chrome_pos selection dominance</span></span>
<span><span class="co">##         &lt;int&gt;  &lt;int&gt;         &lt;int&gt; &lt;chr&gt;         &lt;int&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span></span>
<span><span class="co">##  1       1000     32          1059 m1            51819         0       0.5</span></span>
<span><span class="co">##  2       1000      2          1625 m1            38952         0       0.5</span></span>
<span><span class="co">##  3       1000      7          1785 m1            70816         0       0.5</span></span>
<span><span class="co">##  4       1000     34          1792 m1            65569         0       0.5</span></span>
<span><span class="co">##  5       1000     33          1974 m1            55913         0       0.5</span></span>
<span><span class="co">##  6       1000      1          1994 m1            38596         0       0.5</span></span>
<span><span class="co">##  7       1000     27          2163 m1            11533         0       0.5</span></span>
<span><span class="co">##  8       1000      3          2301 m1            58311         0       0.5</span></span>
<span><span class="co">##  9       1000     35          2947 m1            71275         0       0.5</span></span>
<span><span class="co">## 10       1000     31          2990 m1            46828         0       0.5</span></span>
<span><span class="co">## # ℹ 62 more rows</span></span>
<span><span class="co">## # ℹ 4 more variables: subpop &lt;chr&gt;, first_gen &lt;int&gt;, prevalence &lt;int&gt;,</span></span>
<span><span class="co">## #   nucleotide &lt;chr&gt;</span></span></code></pre>
<p>Okay, so now we are ready to make our callback function. First we
will create a little function to plot the Hilbert Curve. This uses a
trick from the <code>grDevices</code> package to hold the display until
each plot is finished drawing. This gives the animation produced while
running the simulation a smoother appearance, but is not strictly
necessary.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## specify chromosome ranges for plotting</span></span>
<span><span class="va">chrs</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">87382</span>, <span class="fl">174763</span><span class="op">)</span>,</span>
<span>                         end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">87382</span>, <span class="fl">174763</span>, <span class="fl">262143</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bg</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fl">0</span>, end <span class="op">=</span> <span class="fl">262143</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_HC</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">positions</span>, <span class="va">prevalence</span>, <span class="va">chrs</span>, <span class="va">bg</span>, <span class="va">max_prev</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">prevalence</span> <span class="op">&lt;-</span> <span class="va">prevalence</span> <span class="op">/</span> <span class="va">max_prev</span></span>
<span>  <span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">colourvalues</span><span class="fu">::</span><span class="fu"><a href="https://symbolixau.github.io/colourvalues/reference/colour_values.html" class="external-link">color_values</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">prevalence</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">prevalence</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="fu">grDevices</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.flush.html" class="external-link">dev.hold</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HilbertCurve/man/HilbertCurve.html" class="external-link">HilbertCurve</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">262143</span>, level <span class="op">=</span> <span class="fl">4</span>, reference <span class="op">=</span> <span class="cn">TRUE</span>, arrow <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/HilbertCurve/man/hc_polygon-dispatch.html" class="external-link">hc_polygon</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">chrs</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/HilbertCurve/man/hc_points-dispatch.html" class="external-link">hc_points</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">bg</span>, np <span class="op">=</span> <span class="fl">4</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu">colourvalues</span><span class="fu">::</span><span class="fu"><a href="https://symbolixau.github.io/colourvalues/reference/colour_values.html" class="external-link">color_values</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/HilbertCurve/man/hc_points-dispatch.html" class="external-link">hc_points</a></span><span class="op">(</span><span class="va">hc</span>, x1 <span class="op">=</span> <span class="va">positions</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">cols</span><span class="op">)</span>, np <span class="op">=</span> <span class="fl">4</span>, mean_mode <span class="op">=</span> <span class="st">"weighted"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu">grDevices</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.flush.html" class="external-link">dev.flush</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">#grDevices::dev.new(noRStudioGD = TRUE)</span></span>
<span><span class="fu">plot_HC</span><span class="op">(</span><span class="va">mut_dat</span><span class="op">$</span><span class="va">chrome_pos</span>, <span class="va">mut_dat</span><span class="op">$</span><span class="va">prevalence</span>, <span class="va">chrs</span>, <span class="va">bg</span>, max_prev <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p><img src="Custom_vis_files/figure-html/make_callback-1.png" width="700"></p>
<p>Now the final callback function:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">make_hilbert</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">data</span>, <span class="va">max_prev</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">mutation_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim_extract_full.html">slim_extract_full</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, <span class="op">]</span>, <span class="st">"mutations"</span><span class="op">)</span></span>
<span>  <span class="fu">plot_HC</span><span class="op">(</span><span class="va">mutation_data</span><span class="op">$</span><span class="va">chrome_pos</span>, <span class="va">mutation_data</span><span class="op">$</span><span class="va">prevalence</span>, <span class="va">chrs</span>, <span class="va">bg</span>, <span class="va">max_prev</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Let’s try it out!</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## change mutation rate and increase pop size</span></span>
<span><span class="va">chromosome_sim_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim_script_render.html">slim_script_render</a></span><span class="op">(</span><span class="va">chromosome_sim</span>,</span>
<span>                                           template <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                                           mut_rate <span class="op">=</span> <span class="fl">1e-6</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## increase generations</span></span>
<span><span class="fu"><a href="../reference/end_gen.html">end_gen</a></span><span class="op">(</span><span class="va">chromosome_sim_1</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span></span>
<span><span class="co">## open a non-rstudio graphics device (faster and support holding)</span></span>
<span><span class="fu">grDevices</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.new</a></span><span class="op">(</span>noRStudioGD <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">sim_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim_run.html">slim_run</a></span><span class="op">(</span><span class="va">chromosome_sim_1</span>, </span>
<span>                        callbacks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">make_hilbert</span><span class="op">)</span>, </span>
<span>                        cb_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>max_prev <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/HC_callback_vis_demo.gif"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Russell Dinnage, Stephen Sarre, Bernd Gruber, Richard Duncan.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
