<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>simple_example_using_migration_and_fst • slimr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="simple_example_using_migration_and_fst">
<meta property="og:description" content="slimr">
<meta property="og:image" content="https://rdinnager.github.io/slimr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">slimr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Custom_vis.html">Custom Live Plotting for SLiM</a>
    </li>
    <li>
      <a href="../articles/Main_manuscript_example_v2.html">Using slimr to investigate the population genetics of the Sandy Inland Mouse in the periodic rainfall environment of the Simpson Desert</a>
    </li>
    <li>
      <a href="../articles/simple_example_using_migration_and_fst.html">simple_example_using_migration_and_fst</a>
    </li>
    <li>
      <a href="../articles/installing_SLiM_for_slimr.html">How to Install SLiM for Use with slimr</a>
    </li>
    <li>
      <a href="../articles/simple_nucleotide_example.html">Simple Example of a Simulation using Nucleotides</a>
    </li>
    <li>
      <a href="../articles/slimrlang_intro.html">Introduction to slimr: Writing SLiM scripts from R</a>
    </li>
    <li>
      <a href="../articles/useful_tables.html">Useful slimr Tables</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rdinnager/slimr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>simple_example_using_migration_and_fst</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/rdinnager/slimr/blob/HEAD/vignettes/articles/simple_example_using_migration_and_fst.Rmd" class="external-link"><code>vignettes/articles/simple_example_using_migration_and_fst.Rmd</code></a></small>
      <div class="hidden name"><code>simple_example_using_migration_and_fst.Rmd</code></div>

    </div>

    
    
<style type="text/css" scoped>
PRE.fansi SPAN {padding-top: .25em; padding-bottom: .25em};
</style>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rdinnager.github.io/slimr">slimr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Welcome to the slimr package for forward population genetics simulation in SLiM. For more information on SLiM please visit https://messerlab.org/slim/ .</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'slimr'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:methods':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     initialize</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     interaction</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thibautjombart/adegenet" class="external-link">adegenet</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ade4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    /// adegenet 2.1.10 is loaded ////////////</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    &gt; overview: '?adegenet'</span></span>
<span><span class="co">#&gt;    &gt; tutorials/doc/questions: 'adegenetWeb()' </span></span>
<span><span class="co">#&gt;    &gt; bug reports/feature requests: adegenetIssues()</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://green-striped-gecko.github.io/dartR/" class="external-link">dartR</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="co">#&gt; Loading required package: dplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:slimr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     first</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="co">#&gt; Loading required package: dartR.data</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'pegas':</span></span>
<span><span class="co">#&gt;   method      from</span></span>
<span><span class="co">#&gt;   print.amova ade4</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">#&gt;   method from   </span></span>
<span><span class="co">#&gt;   +.gg   ggplot2</span></span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,</span></span>
<span><span class="co">#&gt; which was just loaded, will retire in October 2023.</span></span>
<span><span class="co">#&gt; Please refer to R-spatial evolution reports for details, especially</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html.</span></span>
<span><span class="co">#&gt; It may be desirable to make the sf package available;</span></span>
<span><span class="co">#&gt; package maintainers should consider adding sf to Suggests:.</span></span>
<span><span class="co">#&gt; The sp package is now running under evolution status 2</span></span>
<span><span class="co">#&gt;      (status 2 uses the sf package in place of rgdal)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'genetics':</span></span>
<span><span class="co">#&gt;   method      from </span></span>
<span><span class="co">#&gt;   [.haplotype pegas</span></span>
<span><span class="co">#&gt; Please note that rgdal will be retired during October 2023,</span></span>
<span><span class="co">#&gt; plan transition to sf/stars/terra functions using GDAL and PROJ</span></span>
<span><span class="co">#&gt; at your earliest convenience.</span></span>
<span><span class="co">#&gt; See https://r-spatial.org/r/2023/05/15/evolution4.html and https://github.com/r-spatial/evolution</span></span>
<span><span class="co">#&gt; rgdal: version: 1.6-7, (SVN revision 1203)</span></span>
<span><span class="co">#&gt; Geospatial Data Abstraction Library extensions to R successfully loaded</span></span>
<span><span class="co">#&gt; Loaded GDAL runtime: GDAL 3.6.2, released 2023/01/02</span></span>
<span><span class="co">#&gt; Path to GDAL shared files: C:/R/R-4.3.1/library/rgdal/gdal</span></span>
<span><span class="co">#&gt;  GDAL does not use iconv for recoding strings.</span></span>
<span><span class="co">#&gt; GDAL binary built with GEOS: TRUE </span></span>
<span><span class="co">#&gt; Loaded PROJ runtime: Rel. 9.2.0, March 1st, 2023, [PJ_VERSION: 920]</span></span>
<span><span class="co">#&gt; Path to PROJ shared files: C:/R/R-4.3.1/library/rgdal/proj</span></span>
<span><span class="co">#&gt; PROJ CDN enabled: FALSE</span></span>
<span><span class="co">#&gt; Linking to sp version:2.0-0</span></span>
<span><span class="co">#&gt; To mute warnings of possible GDAL/OSR exportToProj4() degradation,</span></span>
<span><span class="co">#&gt; use options("rgdal_show_exportToProj4_warnings"="none") before loading sp or rgdal.</span></span></code></pre></div>
<pre class="fansi fansi-message"><code><span><span class="co">#&gt; **** Welcome to dartR [Version 2.9.7 ] ****</span></span>
<span><span class="co">#&gt; </span></span></code></pre>
<pre class="fansi fansi-message"><code><span><span class="co">#&gt; Be aware that owing to CRAN requirements and compatibility reasons not all functions of the package may run after the basic installation, as some packages could still be missing. Hence for a most enjoyable experience we recommend to run the function </span></span></code></pre>
<pre class="fansi fansi-message"><code><span><span class="co">#&gt; gl.install.vanilla.dartR()</span></span></code></pre>
<pre class="fansi fansi-message"><code><span><span class="co">#&gt; This installs all missing and required packages for your version of dartR. In case something fails during installation please refer to this tutorial: https://github.com/green-striped-gecko/dartR/wiki/Installation-tutorial.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; For information how to cite dartR, please use:</span></span></code></pre>
<pre class="fansi fansi-message"><code><span><span class="co">#&gt; citation('dartR')</span></span></code></pre>
<pre class="fansi fansi-message"><code><span><span class="co">#&gt; Global verbosity is set to: 2</span></span>
<span><span class="co">#&gt; </span></span></code></pre>
<pre class="fansi fansi-message"><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; **** Have fun using dartR! ****</span></span></code></pre>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/DavisVaughan/furrr" class="external-link">furrr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: future</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123456</span><span class="op">)</span></span></code></pre></div>
<pre><code>initialize() {
  initializeMutationRate(1e-7);
  initializeMutationType("m1", 0.5, "f", 0.0);
  initializeGenomicElementType("g1", m1, 1.0);
  initializeGenomicElement(g1, 0, 99999);
  initializeRecombinationRate(1e-8);
}

1 {
  for (i in 1:3)
    sim.addSubpop(i, 1000);
  subpops = sim.subpopulations;
  lines = readFile("~/Desktop/migration.csv");
  lines = lines[substr(lines, 0, 1) != "//"];
  for (line in lines)
  {
    fields = strsplit(line, ",");
    i = asInteger(fields[0]);
    j = asInteger(fields[1]);
    m = asFloat(fields[2]);
    if (i != j)
    {
      p_i = subpops[subpops.id == i];
      p_j = subpops[subpops.id == j];
      p_j.setMigrationRates(p_i, m);
    }
  }
}

10000 late() { sim.outputFull(); }
</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/slim_script.html">slim_script</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/slim_block.html">slim_block</a></span><span class="op">(</span><span class="fu"><a href="../reference/callback_initialize.html">initialize</a></span><span class="op">(</span><span class="op">)</span>, <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/initializeMutationRate.html">initializeMutationRate</a></span><span class="op">(</span><span class="fl">1e-7</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeMutationType.html">initializeMutationType</a></span><span class="op">(</span><span class="st">"m1"</span>, <span class="fl">0.5</span>, <span class="st">"f"</span>, <span class="fl">0.0</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeGenomicElementType.html">initializeGenomicElementType</a></span><span class="op">(</span><span class="st">"g1"</span>, <span class="va">m1</span>, <span class="fl">1.0</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeGenomicElement.html">initializeGenomicElement</a></span><span class="op">(</span><span class="va">g1</span>, <span class="fl">0</span>, <span class="fl">99999</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeRecombinationRate.html">initializeRecombinationRate</a></span><span class="op">(</span><span class="fl">1e-8</span><span class="op">)</span>;</span>
<span>  <span class="op">}</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/slim_block.html">slim_block</a></span><span class="op">(</span><span class="fl">1</span>, <span class="op">{</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="fu">sim.addSubpop</span><span class="op">(</span><span class="va">i</span>, <span class="fl">1000</span><span class="op">)</span>;</span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">subpops</span> <span class="op">=</span> <span class="va">sim.subpopulations</span>;</span>
<span>    <span class="va">lines</span> <span class="op">=</span> <span class="fu">readFile</span><span class="op">(</span><span class="st">"~/Desktop/migration.csv"</span><span class="op">)</span>;</span>
<span>    <span class="va">lines</span> <span class="op">=</span> <span class="va">lines</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">lines</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span> <span class="op">!=</span> <span class="st">"//"</span><span class="op">]</span>;</span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">line</span> <span class="kw">in</span> <span class="va">lines</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="va">fields</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">line</span>, <span class="st">","</span><span class="op">)</span>;</span>
<span>      <span class="va">i</span> <span class="op">=</span> <span class="fu">asInteger</span><span class="op">(</span><span class="va">fields</span><span class="op">[</span><span class="fl">0</span><span class="op">]</span><span class="op">)</span>;</span>
<span>      <span class="va">j</span> <span class="op">=</span> <span class="fu">asInteger</span><span class="op">(</span><span class="va">fields</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>;</span>
<span>      <span class="va">m</span> <span class="op">=</span> <span class="fu">asFloat</span><span class="op">(</span><span class="va">fields</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>;</span>
<span>      <span class="kw">if</span> <span class="op">(</span><span class="va">i</span> <span class="op">!=</span> <span class="va">j</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">p_i</span> <span class="op">=</span> <span class="va">subpops</span><span class="op">[</span><span class="va">subpops.id</span> <span class="op">==</span> <span class="va">i</span><span class="op">]</span>;</span>
<span>        <span class="va">p_j</span> <span class="op">=</span> <span class="va">subpops</span><span class="op">[</span><span class="va">subpops.id</span> <span class="op">==</span> <span class="va">j</span><span class="op">]</span>;</span>
<span>        <span class="fu">p_j.setMigrationRates</span><span class="op">(</span><span class="va">p_i</span>, <span class="va">m</span><span class="op">)</span>;</span>
<span>      <span class="op">}</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/slim_block.html">slim_block</a></span><span class="op">(</span><span class="fl">10000</span>, <span class="fu"><a href="../reference/late.html">late</a></span><span class="op">(</span><span class="op">)</span>, <span class="op">{</span></span>
<span>    <span class="fu">sim.outputFull</span><span class="op">(</span><span class="op">)</span>;</span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">script_1</span></span>
<span></span>
<span><span class="va">script_1</span></span></code></pre></div>
<pre class="fansi fansi-output"><code><span><span class="co">#&gt; &lt;slimr_script[3]&gt;</span></span>
<span><span class="co">#&gt; block_init:initialize() {</span></span>
<span><span class="co">#&gt;     initializeMutationRate(1e-07);</span></span>
<span><span class="co">#&gt;     initializeMutationType("m1", 0.5, "f", 0);</span></span>
<span><span class="co">#&gt;     initializeGenomicElementType("g1", m1, 1);</span></span>
<span><span class="co">#&gt;     initializeGenomicElement(g1, 0, 99999);</span></span>
<span><span class="co">#&gt;     initializeRecombinationRate(1e-08);</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_2:1 early() {</span></span>
<span><span class="co">#&gt;     for (i in 1:3) {</span></span>
<span><span class="co">#&gt;         sim.addSubpop(i, 1000);</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt;     subpops = sim.subpopulations;</span></span>
<span><span class="co">#&gt;     lines = readFile("~/Desktop/migration.csv");</span></span>
<span><span class="co">#&gt;     lines = lines[substr(lines, 0, 1) != "//"];</span></span>
<span><span class="co">#&gt;     for (line in lines) {</span></span>
<span><span class="co">#&gt;         fields = strsplit(line, ",");</span></span>
<span><span class="co">#&gt;         i = asInteger(fields[0]);</span></span>
<span><span class="co">#&gt;         j = asInteger(fields[1]);</span></span>
<span><span class="co">#&gt;         m = asFloat(fields[2]);</span></span>
<span><span class="co">#&gt;         if (i != j) {</span></span>
<span><span class="co">#&gt;             p_i = subpops[subpops.id == i];</span></span>
<span><span class="co">#&gt;             p_j = subpops[subpops.id == j];</span></span>
<span><span class="co">#&gt;             p_j.setMigrationRates(p_i, m);</span></span>
<span><span class="co">#&gt;         }</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_3:10000 late() {</span></span>
<span><span class="co">#&gt;     sim.outputFull();</span></span>
<span><span class="co">#&gt; }</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">disp_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0.78</span>, </span>
<span>                     <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">0.10</span>, </span>
<span>                     <span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">0.12</span>, </span>
<span>                     <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">0.01</span>, </span>
<span>                     <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">0.96</span>, </span>
<span>                     <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">0.03</span>, </span>
<span>                     <span class="fl">3</span>, <span class="fl">1</span>, <span class="fl">0.33</span>, </span>
<span>                     <span class="fl">3</span>, <span class="fl">2</span>, <span class="fl">0.17</span>, </span>
<span>                     <span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>                   ncol <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/slim_script.html">slim_script</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/slim_block.html">slim_block</a></span><span class="op">(</span><span class="fu"><a href="../reference/callback_initialize.html">initialize</a></span><span class="op">(</span><span class="op">)</span>, <span class="op">{</span></span>
<span>    <span class="fu">setSeed</span><span class="op">(</span><span class="fl">123456</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeMutationRate.html">initializeMutationRate</a></span><span class="op">(</span><span class="fl">2e-6</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeMutationType.html">initializeMutationType</a></span><span class="op">(</span><span class="st">"m1"</span>, <span class="fl">0.5</span>, <span class="st">"f"</span>, <span class="fl">0.0</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeGenomicElementType.html">initializeGenomicElementType</a></span><span class="op">(</span><span class="st">"g1"</span>, <span class="va">m1</span>, <span class="fl">1.0</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeGenomicElement.html">initializeGenomicElement</a></span><span class="op">(</span><span class="va">g1</span>, <span class="fl">0</span>, <span class="fl">99999</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeRecombinationRate.html">initializeRecombinationRate</a></span><span class="op">(</span><span class="fl">1e-8</span><span class="op">)</span>;</span>
<span>  <span class="op">}</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/slim_block.html">slim_block</a></span><span class="op">(</span><span class="fl">1</span>, <span class="op">{</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="fu">sim.addSubpop</span><span class="op">(</span><span class="va">i</span>, <span class="fl">100</span><span class="op">)</span>;</span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">subpops</span> <span class="op">=</span> <span class="va">sim.subpopulations</span>;</span>
<span>    </span>
<span>    <span class="va">disp_mat</span> <span class="op">=</span> <span class="fu"><a href="../reference/r_inline.html">slimr_inline</a></span><span class="op">(</span><span class="va">disp_mat</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">line</span> <span class="kw">in</span> <span class="fu">seqLen</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">disp_mat</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="va">i</span> <span class="op">=</span> <span class="fu">asInteger</span><span class="op">(</span><span class="va">disp_mat</span><span class="op">[</span><span class="va">line</span>, <span class="fl">0</span><span class="op">]</span><span class="op">)</span>;</span>
<span>      <span class="va">j</span> <span class="op">=</span> <span class="fu">asInteger</span><span class="op">(</span><span class="va">disp_mat</span><span class="op">[</span><span class="va">line</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>;</span>
<span>      <span class="va">m</span> <span class="op">=</span> <span class="fu">asFloat</span><span class="op">(</span><span class="va">disp_mat</span><span class="op">[</span><span class="va">line</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span>;</span>
<span>      <span class="kw">if</span> <span class="op">(</span><span class="va">i</span> <span class="op">!=</span> <span class="va">j</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">p_i</span> <span class="op">=</span> <span class="va">subpops</span><span class="op">[</span><span class="va">subpops.id</span> <span class="op">==</span> <span class="va">i</span><span class="op">]</span>;</span>
<span>        <span class="va">p_j</span> <span class="op">=</span> <span class="va">subpops</span><span class="op">[</span><span class="va">subpops.id</span> <span class="op">==</span> <span class="va">j</span><span class="op">]</span>;</span>
<span>        <span class="fu">p_j.setMigrationRates</span><span class="op">(</span><span class="va">p_i</span>, <span class="va">m</span><span class="op">)</span>;</span>
<span>      <span class="op">}</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/slim_block.html">slim_block</a></span><span class="op">(</span><span class="fl">100000</span>, <span class="fu"><a href="../reference/late.html">late</a></span><span class="op">(</span><span class="op">)</span>, <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/r_output.html">slimr_output</a></span><span class="op">(</span><span class="fu">sim.outputFull</span><span class="op">(</span><span class="op">)</span>, <span class="st">"final_output"</span><span class="op">)</span>;</span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">script_2</span></span>
<span></span>
<span><span class="va">script_2</span></span></code></pre></div>
<pre class="fansi fansi-output"><code><span><span class="co">#&gt; &lt;slimr_script[3]&gt;</span></span>
<span><span class="co">#&gt; block_init:initialize() {</span></span>
<span><span class="co">#&gt;     setSeed(123456);</span></span>
<span><span class="co">#&gt;     initializeMutationRate(2e-06);</span></span>
<span><span class="co">#&gt;     initializeMutationType("m1", 0.5, "f", 0);</span></span>
<span><span class="co">#&gt;     initializeGenomicElementType("g1", m1, 1);</span></span>
<span><span class="co">#&gt;     initializeGenomicElement(g1, 0, 99999);</span></span>
<span><span class="co">#&gt;     initializeRecombinationRate(1e-08);</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_2:1 early() {</span></span>
<span><span class="co">#&gt;     for (i in 1:3) {</span></span>
<span><span class="co">#&gt;         sim.addSubpop(i, 100);</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt;     subpops = sim.subpopulations;</span></span>
<span><span class="co">#&gt;     disp_mat = {. = matrix(c(1, 1, 1, 2, 2, 2, 3, 3, 3, 1, ...), nrow = 9, ncol = 3)}</span></span>
<span><span class="co">#&gt;     for (line in seqLen(nrow(disp_mat))) {</span></span>
<span><span class="co">#&gt;         i = asInteger(disp_mat[line, 0]);</span></span>
<span><span class="co">#&gt;         j = asInteger(disp_mat[line, 1]);</span></span>
<span><span class="co">#&gt;         m = asFloat(disp_mat[line, 2]);</span></span>
<span><span class="co">#&gt;         if (i != j) {</span></span>
<span><span class="co">#&gt;             p_i = subpops[subpops.id == i];</span></span>
<span><span class="co">#&gt;             p_j = subpops[subpops.id == j];</span></span>
<span><span class="co">#&gt;             p_j.setMigrationRates(p_i, m);</span></span>
<span><span class="co">#&gt;         }</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_3:1e+05 late() {</span></span>
<span><span class="co">#&gt;     {sim.outputFull() -&gt; final_output}</span></span>
<span><span class="co">#&gt; }</span></span></code></pre>
<p>We can see that the whole matrix has been embedded in the SLiM script
by running <code>as_slim_text</code> on this <code>slimr_script</code>
object. This is the script that will actually be run by SLiM.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="../reference/as_slim_text.html">as_slim_text</a></span><span class="op">(</span><span class="va">script_2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; initialize() {</span></span>
<span><span class="co">#&gt;     setSeed(123456);</span></span>
<span><span class="co">#&gt;     initializeMutationRate(2e-06);</span></span>
<span><span class="co">#&gt;     initializeMutationType("m1", 0.5, "f", 0);</span></span>
<span><span class="co">#&gt;     initializeGenomicElementType("g1", m1, 1);</span></span>
<span><span class="co">#&gt;     initializeGenomicElement(g1, 0, 99999);</span></span>
<span><span class="co">#&gt;     initializeRecombinationRate(1e-08);</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1 early() {</span></span>
<span><span class="co">#&gt;     for (i in 1:3) {</span></span>
<span><span class="co">#&gt;         sim.addSubpop(i, 100);</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt;     subpops = sim.subpopulations;</span></span>
<span><span class="co">#&gt;     disp_mat = matrix(c(1, 1, 1, 2, 2, 2, 3, 3, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 0.78, 0.1, 0.12, 0.01, 0.96, 0.03, 0.33, 0.17, 0.5), nrow = 9, ncol = 3);</span></span>
<span><span class="co">#&gt;     for (line in seqLen(nrow(disp_mat))) {</span></span>
<span><span class="co">#&gt;         i = asInteger(disp_mat[line, 0]);</span></span>
<span><span class="co">#&gt;         j = asInteger(disp_mat[line, 1]);</span></span>
<span><span class="co">#&gt;         m = asFloat(disp_mat[line, 2]);</span></span>
<span><span class="co">#&gt;         if (i != j) {</span></span>
<span><span class="co">#&gt;             p_i = subpops[subpops.id == i];</span></span>
<span><span class="co">#&gt;             p_j = subpops[subpops.id == j];</span></span>
<span><span class="co">#&gt;             p_j.setMigrationRates(p_i, m);</span></span>
<span><span class="co">#&gt;         }</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1e+05 late() {</span></span>
<span><span class="co">#&gt;     if (community.tick%1 == 0) {</span></span>
<span><span class="co">#&gt;         cat("\n&lt;slimr_out:start&gt;\n" + paste(community.tick) + ",'" + "final_output" + "','" + "sim.outputFull()" + "','" + "slim_output','");</span></span>
<span><span class="co">#&gt;         sim.outputFull();</span></span>
<span><span class="co">#&gt;         cat("'\n&lt;slimr_out:end&gt;\n");</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt; }</span></span></code></pre></div>
<p>Now let’s see if we can run it.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim_run.html">slim_run</a></span><span class="op">(</span><span class="va">script_2</span>, throw_error <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Simulation finished with exit status: 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Success!</span></span></code></pre></div>
<p>Looks like it worked. We can get the resulting genomic data from the
returned object and convert it into a genlight object using the
<code><a href="../reference/slim_extract_genlight.html">slim_extract_genlight()</a></code> function. Then we can plot the
results.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://green-striped-gecko.github.io/dartR/" class="external-link">dartR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim_extract_genlight.html">slim_extract_genlight</a></span><span class="op">(</span><span class="va">results</span>, <span class="st">"final_output"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">ploidy</a></span><span class="op">(</span><span class="va">inds</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">inds</span><span class="op">)</span></span></code></pre></div>
<p><img src="simple_example_using_migration_and_fst_files/figure-html/get_dat-1.png" width="700">
Note that the SNPs are ordered by their mutation id, which should
roughly correspond to their order of origination, so the leftmost
mutations are the oldest, the rightmost the youngest. Using a genlight
object we can calculate a lot of population genetics metrics easily
using the package <code>dartR</code>. Let’s calculate Fst for
example.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_fst</span> <span class="op">&lt;-</span> <span class="fu">dartR</span><span class="fu">::</span><span class="fu"><a href="https://green-striped-gecko.github.io/dartR/reference/gl.fst.pop.html" class="external-link">gl.fst.pop</a></span><span class="op">(</span><span class="va">inds</span>, nboots <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">p_fst</span></span>
<span><span class="co">#&gt;              p2         p3 p1</span></span>
<span><span class="co">#&gt; p2           NA         NA NA</span></span>
<span><span class="co">#&gt; p3 0.0005873432         NA NA</span></span>
<span><span class="co">#&gt; p1 0.0139928734 0.00888913 NA</span></span></code></pre></div>
<p>So generally those Fst values are low, which we would expect since
our populations are not experiencing differential selection and the
effects of drift are counteracted by migration between the populations.
Nevertheless the pairwise Fst’s measured here are inversely proportional
to the overall strength of migration between the population in our
simulation. We can see this like this:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## convert migration rates into pairwise matrix using igraph</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/" class="external-link">igraph</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'igraph'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:future':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     %-&gt;%, %&lt;-%</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:purrr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     compose, simplify</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     as_data_frame, groups, union</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     decompose, spectrum</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     union</span></span>
<span><span class="va">overall_mig_rates</span> <span class="op">&lt;-</span> <span class="va">disp_mat</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/graph_from_data_frame.html" class="external-link">graph_from_data_frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/as.directed.html" class="external-link">as.undirected</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"collapse"</span>, edge.attr.comb <span class="op">=</span> <span class="st">"mean"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/as_adjacency_matrix.html" class="external-link">as_adjacency_matrix</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"lower"</span>, attr <span class="op">=</span> <span class="st">"V3"</span>, sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">overall_mig_rates</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">overall_mig_rates</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lower.tri.html" class="external-link">upper.tri</a></span><span class="op">(</span><span class="va">overall_mig_rates</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span></span>
<span><span class="va">p_fst</span></span>
<span><span class="co">#&gt;              p2         p3 p1</span></span>
<span><span class="co">#&gt; p2           NA         NA NA</span></span>
<span><span class="co">#&gt; p3 0.0005873432         NA NA</span></span>
<span><span class="co">#&gt; p1 0.0139928734 0.00888913 NA</span></span>
<span><span class="va">overall_mig_rates</span></span>
<span><span class="co">#&gt;       1   2  3</span></span>
<span><span class="co">#&gt; 1    NA  NA NA</span></span>
<span><span class="co">#&gt; 2 0.055  NA NA</span></span>
<span><span class="co">#&gt; 3 0.225 0.1 NA</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">overall_mig_rates</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">p_fst</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="simple_example_using_migration_and_fst_files/figure-html/compare_mig-1.png" width="700"></p>
<p>We can see generally the populations with the lowest migration rates
also has the highest pairwise Fst, suggesting population with low
migration have differentiated the most, which makes sense. In this
particular example, the Fst seems to reach a minimum soon after, such
that the two higher migration pairs don’t differ much in their pairwise
Fsts. This could be a result of stochasticity. if we really want to know
whether migration rate effects Fst under a simple neutral model of
evolution, we should run many simulations and look at the average
effect. <code>slimr</code> makes this easy to do using templating. Let’s
simplify this model a little to just look at two populations at a time,
and then set it up to run many times with different migration rates
between them. Ww can also vary population size as it seems like this
should be an important mediator of any effects, given it will effect the
relative strength of drift in the populations.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/slim_script.html">slim_script</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/slim_block.html">slim_block</a></span><span class="op">(</span><span class="fu"><a href="../reference/callback_initialize.html">initialize</a></span><span class="op">(</span><span class="op">)</span>, <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/initializeMutationRate.html">initializeMutationRate</a></span><span class="op">(</span><span class="fl">2e-6</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeMutationType.html">initializeMutationType</a></span><span class="op">(</span><span class="st">"m1"</span>, <span class="fl">0.5</span>, <span class="st">"f"</span>, <span class="fl">0.0</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeGenomicElementType.html">initializeGenomicElementType</a></span><span class="op">(</span><span class="st">"g1"</span>, <span class="va">m1</span>, <span class="fl">1.0</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeGenomicElement.html">initializeGenomicElement</a></span><span class="op">(</span><span class="va">g1</span>, <span class="fl">0</span>, <span class="fl">99999</span><span class="op">)</span>;</span>
<span>    <span class="fu"><a href="../reference/initializeRecombinationRate.html">initializeRecombinationRate</a></span><span class="op">(</span><span class="fl">1e-8</span><span class="op">)</span>;</span>
<span>  <span class="op">}</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/slim_block.html">slim_block</a></span><span class="op">(</span><span class="fl">1</span>, <span class="op">{</span></span>
<span>    <span class="fu">sim.addSubpop</span><span class="op">(</span><span class="st">"p1"</span>, <span class="fu"><a href="../reference/r_template.html">slimr_template</a></span><span class="op">(</span><span class="st">"pop_size"</span>, default <span class="op">=</span> <span class="fl">100</span><span class="op">)</span><span class="op">)</span>;</span>
<span>    <span class="fu">sim.addSubpop</span><span class="op">(</span><span class="st">"p2"</span>, <span class="va">..pop_size..</span><span class="op">)</span>;</span>
<span>    <span class="fu">p1.setMigrationRates</span><span class="op">(</span><span class="va">p2</span>, <span class="fu"><a href="../reference/r_template.html">slimr_template</a></span><span class="op">(</span><span class="st">"m"</span>, default <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span>;</span>
<span>    <span class="fu">p2.setMigrationRates</span><span class="op">(</span><span class="va">p1</span>, <span class="va">..m..</span><span class="op">)</span>;</span>
<span>  <span class="op">}</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/slim_block.html">slim_block</a></span><span class="op">(</span><span class="fl">10000</span>, <span class="fu"><a href="../reference/late.html">late</a></span><span class="op">(</span><span class="op">)</span>, <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/r_output.html">slimr_output</a></span><span class="op">(</span><span class="fu">sim.outputFull</span><span class="op">(</span><span class="op">)</span>, <span class="st">"final_output"</span><span class="op">)</span>;</span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">script_3</span></span>
<span></span>
<span><span class="va">script_3</span></span></code></pre></div>
<pre class="fansi fansi-output"><code><span><span class="co">#&gt; &lt;slimr_script[3]&gt;</span></span>
<span><span class="co">#&gt; block_init:initialize() {</span></span>
<span><span class="co">#&gt;     initializeMutationRate(2e-06);</span></span>
<span><span class="co">#&gt;     initializeMutationType("m1", 0.5, "f", 0);</span></span>
<span><span class="co">#&gt;     initializeGenomicElementType("g1", m1, 1);</span></span>
<span><span class="co">#&gt;     initializeGenomicElement(g1, 0, 99999);</span></span>
<span><span class="co">#&gt;     initializeRecombinationRate(1e-08);</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_2:1 early() {</span></span>
<span><span class="co">#&gt;     sim.addSubpop("p1", ..pop_size..);</span></span>
<span><span class="co">#&gt;     sim.addSubpop("p2", ..pop_size..);</span></span>
<span><span class="co">#&gt;     p1.setMigrationRates(p2, ..m..);</span></span>
<span><span class="co">#&gt;     p2.setMigrationRates(p1, ..m..);</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_3:10000 late() {</span></span>
<span><span class="co">#&gt;     {sim.outputFull() -&gt; final_output}</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; This slimr_script has templating in block(s) block_2 for</span></span>
<span><span class="co">#&gt; variables pop_size and m.</span></span></code></pre>
<p>So in the above script we now have a placemarker for a
<code>pop_size</code> and a <code>m</code> parameter. We specified this
using the <code>slimr_template</code> function directly inserted into
the SLiM code. We also demonstrate a trick to reduce typing. We only
have to use the <code>slimr_template</code> function once for a given
variable, subsequent references to the same variable can be specified
directly using the <code>..x..</code> notation. Now to generate a script
with the templated variables replaced with values of our choosing, we
use the <code>slim_script_render</code> function. Here is an
example:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">example_script</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim_script_render.html">slim_script_render</a></span><span class="op">(</span><span class="va">script_3</span>, template <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>pop_size <span class="op">=</span> <span class="fl">50</span>, m <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">example_script</span></span></code></pre></div>
<pre class="fansi fansi-output"><code><span><span class="co">#&gt; &lt;slimr_script[3]&gt;</span></span>
<span><span class="co">#&gt; block_init:initialize() {</span></span>
<span><span class="co">#&gt;     initializeMutationRate(2e-06);</span></span>
<span><span class="co">#&gt;     initializeMutationType("m1", 0.5, "f", 0);</span></span>
<span><span class="co">#&gt;     initializeGenomicElementType("g1", m1, 1);</span></span>
<span><span class="co">#&gt;     initializeGenomicElement(g1, 0, 99999);</span></span>
<span><span class="co">#&gt;     initializeRecombinationRate(1e-08);</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_2:1 early() {</span></span>
<span><span class="co">#&gt;     sim.addSubpop("p1", 50);</span></span>
<span><span class="co">#&gt;     sim.addSubpop("p2", 50);</span></span>
<span><span class="co">#&gt;     p1.setMigrationRates(p2, 0.05);</span></span>
<span><span class="co">#&gt;     p2.setMigrationRates(p1, 0.05);</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_3:10000 late() {</span></span>
<span><span class="co">#&gt;     {sim.outputFull() -&gt; final_output}</span></span>
<span><span class="co">#&gt; }</span></span></code></pre>
<p>We can also pass a <code>data.frame</code> to
<code>slim_script_render</code> which will allow us to generate many
scripts at once, let’s setup a large number of scripts to explore
different values of our two parameters.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_reps</span> <span class="op">&lt;-</span> <span class="fl">300</span></span>
<span></span>
<span><span class="va">param_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>pop_size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">5</span><span class="op">:</span><span class="fl">100</span>, <span class="va">n_reps</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                                                         m <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="va">n_reps</span>, <span class="fl">0</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">script_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim_script_render.html">slim_script_render</a></span><span class="op">(</span><span class="va">script_3</span>,</span>
<span>                                   template <span class="op">=</span> <span class="va">param_df</span><span class="op">)</span></span>
<span></span>
<span><span class="va">script_list</span></span></code></pre></div>
<pre class="fansi fansi-output"><code><span><span class="co">#&gt; &lt;slimr_script_coll[300]&gt;</span></span>
<span><span class="co">#&gt; &lt;1&gt;</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_init:initialize() {</span></span>
<span><span class="co">#&gt;     initializeMutationRate(2e-06);</span></span>
<span><span class="co">#&gt;     initializeMutationType("m1", 0.5, "f", 0);</span></span>
<span><span class="co">#&gt;     initializeGenomicElementType("g1", m1, 1);</span></span>
<span><span class="co">#&gt;     initializeGenomicElement(g1, 0, 99999);</span></span>
<span><span class="co">#&gt;     initializeRecombinationRate(1e-08);</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_2:1 early() {</span></span>
<span><span class="co">#&gt;     sim.addSubpop("p1", 85);</span></span>
<span><span class="co">#&gt;     sim.addSubpop("p2", 85);</span></span>
<span><span class="co">#&gt;     p1.setMigrationRates(p2, 0.379254172206856);</span></span>
<span><span class="co">#&gt;     p2.setMigrationRates(p1, 0.379254172206856);</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_3:10000 late() {</span></span>
<span><span class="co">#&gt;     {sim.outputFull() -&gt; final_output}</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; &lt;2&gt;</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_init:initialize() {</span></span>
<span><span class="co">#&gt;     initializeMutationRate(2e-06);</span></span>
<span><span class="co">#&gt;     initializeMutationType("m1", 0.5, "f", 0);</span></span>
<span><span class="co">#&gt;     initializeGenomicElementType("g1", m1, 1);</span></span>
<span><span class="co">#&gt;     initializeGenomicElement(g1, 0, 99999);</span></span>
<span><span class="co">#&gt;     initializeRecombinationRate(1e-08);</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_2:1 early() {</span></span>
<span><span class="co">#&gt;     sim.addSubpop("p1", 41);</span></span>
<span><span class="co">#&gt;     sim.addSubpop("p2", 41);</span></span>
<span><span class="co">#&gt;     p1.setMigrationRates(p2, 0.26800632593222);</span></span>
<span><span class="co">#&gt;     p2.setMigrationRates(p1, 0.26800632593222);</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_3:10000 late() {</span></span>
<span><span class="co">#&gt;     {sim.outputFull() -&gt; final_output}</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; &lt;3&gt;</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_init:initialize() {</span></span>
<span><span class="co">#&gt;     initializeMutationRate(2e-06);</span></span>
<span><span class="co">#&gt;     initializeMutationType("m1", 0.5, "f", 0);</span></span>
<span><span class="co">#&gt;     initializeGenomicElementType("g1", m1, 1);</span></span>
<span><span class="co">#&gt;     initializeGenomicElement(g1, 0, 99999);</span></span>
<span><span class="co">#&gt;     initializeRecombinationRate(1e-08);</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_2:1 early() {</span></span>
<span><span class="co">#&gt;     sim.addSubpop("p1", 37);</span></span>
<span><span class="co">#&gt;     sim.addSubpop("p2", 37);</span></span>
<span><span class="co">#&gt;     p1.setMigrationRates(p2, 0.0882677737390622);</span></span>
<span><span class="co">#&gt;     p2.setMigrationRates(p1, 0.0882677737390622);</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; block_3:10000 late() {</span></span>
<span><span class="co">#&gt;     {sim.outputFull() -&gt; final_output}</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; &lt;...&gt;</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; and 297 more.</span></span></code></pre>
<p>This function returns a <code>slimr_script_coll</code> object, or a
slimr script collection. We can now use <code>slim_run</code> on this
object, which will run every script in the collection, one after the
other, and collect the results. We can specify a
<code>parallel = TRUE</code> argument to run the scripts in parallel.
<code>slimr</code> uses the <code>future</code> package for
parallelization. This means that the user must specify a future
<code>plan</code> prior to running <code>slim_run</code> which will
setup the workers to be used by <code>slimr</code>. If you want to use
parallelization with <code>slimr</code>, we recommend you familiarize
yourself with the <code>future</code> package. If in doubt
<code>future::plan(future::multisession)</code> is the most general
purpose specification which should work well on all platforms (and is
the recommended plan by future for best stability of parallelization).
You must also ensure the <code>future</code> and <code>furrr</code>
packages are installed for this functionality to work. Let’s try running
these:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">start_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"furrr"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="fu">future</span><span class="fu">::</span><span class="va"><a href="https://future.futureverse.org/reference/multisession.html" class="external-link">multisession</a></span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">all_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim_run.html">slim_run</a></span><span class="op">(</span><span class="va">script_list</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span>, progress <span class="op">=</span> <span class="cn">TRUE</span>, throw_error <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span></span></code></pre></div>
<p>Cool! First let’s check if all simulation completed without
error:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">errors</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_lgl</a></span><span class="op">(</span><span class="va">all_results</span>,</span>
<span>                         <span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">.x</span><span class="op">$</span><span class="va">error</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">errors</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>Let’s extract all the genomic data into genlight objects so we can
easily calculate Fst. Since this might also take a little while we will
parallelize this extraction as well using <code>furrr</code> (since we
already set a <code>future</code> plan, and <code>furrr</code> uses
<code>future</code> under the hood). Just in case any simulations give
us trouble, we will wrap the extraction in <code><a href="https://rdrr.io/r/base/try.html" class="external-link">try()</a></code> so that
the whole thing doesn’t fail for one error.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_gl</span> <span class="op">&lt;-</span> <span class="fu">furrr</span><span class="fu">::</span><span class="fu"><a href="https://furrr.futureverse.org/reference/future_map.html" class="external-link">future_map</a></span><span class="op">(</span><span class="va">all_results</span>,</span>
<span>                     <span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span><span class="fu"><a href="../reference/slim_extract_genlight.html">slim_extract_genlight</a></span><span class="op">(</span><span class="va">.x</span>, <span class="st">"final_output"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     .progress <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Any errors?</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">errored</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">all_gl</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">==</span> <span class="st">"try-error"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">errored</span></span>
<span><span class="co">#&gt; integer(0)</span></span>
<span></span>
<span><span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">errored</span>,</span>
<span>           <span class="op">~</span><span class="va">all_results</span><span class="op">[[</span><span class="va">.x</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">output_data</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="co">#&gt; list()</span></span></code></pre></div>
<p>We got ‘r length(errored)’ errors. Examining their data strings, we
can see that they all essentially had no mutations, which is why we
couldn’t construct a genlight object. Let’s just filter these out and
continue with our analysis by calculating Fst.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">errored</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">all_gl</span> <span class="op">&lt;-</span> <span class="va">all_gl</span><span class="op">[</span><span class="op">-</span><span class="va">errored</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">all_fsts</span> <span class="op">&lt;-</span> <span class="fu">furrr</span><span class="fu">::</span><span class="fu"><a href="https://furrr.futureverse.org/reference/future_map.html" class="external-link">future_map_dbl</a></span><span class="op">(</span><span class="va">all_gl</span>,</span>
<span>                                  <span class="op">~</span> <span class="op">{</span></span>
<span>                                    <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">ploidy</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span>                                    <span class="fu">dartR</span><span class="fu">::</span><span class="fu"><a href="https://green-striped-gecko.github.io/dartR/reference/gl.fst.pop.html" class="external-link">gl.fst.pop</a></span><span class="op">(</span><span class="va">.x</span>, nboots <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">]</span></span>
<span>                                  <span class="op">}</span>,</span>
<span>                                  .progress <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'pegas':</span></span>
<span><span class="co">#&gt;   method      from</span></span>
<span><span class="co">#&gt;   print.amova ade4</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">#&gt;   method from   </span></span>
<span><span class="co">#&gt;   +.gg   ggplot2</span></span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,</span></span>
<span><span class="co">#&gt; which was just loaded, will retire in October 2023.</span></span>
<span><span class="co">#&gt; Please refer to R-spatial evolution reports for details, especially</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html.</span></span>
<span><span class="co">#&gt; It may be desirable to make the sf package available;</span></span>
<span><span class="co">#&gt; package maintainers should consider adding sf to Suggests:.</span></span>
<span><span class="co">#&gt; The sp package is now running under evolution status 2</span></span>
<span><span class="co">#&gt;      (status 2 uses the sf package in place of rgdal)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'genetics':</span></span>
<span><span class="co">#&gt;   method      from </span></span>
<span><span class="co">#&gt;   [.haplotype pegas</span></span>
<span><span class="co">#&gt; Please note that rgdal will be retired during October 2023,</span></span>
<span><span class="co">#&gt; plan transition to sf/stars/terra functions using GDAL and PROJ</span></span>
<span><span class="co">#&gt; at your earliest convenience.</span></span>
<span><span class="co">#&gt; See https://r-spatial.org/r/2023/05/15/evolution4.html and https://github.com/r-spatial/evolution</span></span>
<span><span class="co">#&gt; rgdal: version: 1.6-7, (SVN revision 1203)</span></span>
<span><span class="co">#&gt; Geospatial Data Abstraction Library extensions to R successfully loaded</span></span>
<span><span class="co">#&gt; Loaded GDAL runtime: GDAL 3.6.2, released 2023/01/02</span></span>
<span><span class="co">#&gt; Path to GDAL shared files: C:/R/R-4.3.1/library/rgdal/gdal</span></span>
<span><span class="co">#&gt;  GDAL does not use iconv for recoding strings.</span></span>
<span><span class="co">#&gt; GDAL binary built with GEOS: TRUE </span></span>
<span><span class="co">#&gt; Loaded PROJ runtime: Rel. 9.2.0, March 1st, 2023, [PJ_VERSION: 920]</span></span>
<span><span class="co">#&gt; Path to PROJ shared files: C:/R/R-4.3.1/library/rgdal/proj</span></span>
<span><span class="co">#&gt; PROJ CDN enabled: FALSE</span></span>
<span><span class="co">#&gt; Linking to sp version:2.0-0</span></span>
<span><span class="co">#&gt; To mute warnings of possible GDAL/OSR exportToProj4() degradation,</span></span>
<span><span class="co">#&gt; use options("rgdal_show_exportToProj4_warnings"="none") before loading sp or rgdal.</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'pegas':</span></span>
<span><span class="co">#&gt;   method      from</span></span>
<span><span class="co">#&gt;   print.amova ade4</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">#&gt;   method from   </span></span>
<span><span class="co">#&gt;   +.gg   ggplot2</span></span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,</span></span>
<span><span class="co">#&gt; which was just loaded, will retire in October 2023.</span></span>
<span><span class="co">#&gt; Please refer to R-spatial evolution reports for details, especially</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html.</span></span>
<span><span class="co">#&gt; It may be desirable to make the sf package available;</span></span>
<span><span class="co">#&gt; package maintainers should consider adding sf to Suggests:.</span></span>
<span><span class="co">#&gt; The sp package is now running under evolution status 2</span></span>
<span><span class="co">#&gt;      (status 2 uses the sf package in place of rgdal)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'genetics':</span></span>
<span><span class="co">#&gt;   method      from </span></span>
<span><span class="co">#&gt;   [.haplotype pegas</span></span>
<span><span class="co">#&gt; Please note that rgdal will be retired during October 2023,</span></span>
<span><span class="co">#&gt; plan transition to sf/stars/terra functions using GDAL and PROJ</span></span>
<span><span class="co">#&gt; at your earliest convenience.</span></span>
<span><span class="co">#&gt; See https://r-spatial.org/r/2023/05/15/evolution4.html and https://github.com/r-spatial/evolution</span></span>
<span><span class="co">#&gt; rgdal: version: 1.6-7, (SVN revision 1203)</span></span>
<span><span class="co">#&gt; Geospatial Data Abstraction Library extensions to R successfully loaded</span></span>
<span><span class="co">#&gt; Loaded GDAL runtime: GDAL 3.6.2, released 2023/01/02</span></span>
<span><span class="co">#&gt; Path to GDAL shared files: C:/R/R-4.3.1/library/rgdal/gdal</span></span>
<span><span class="co">#&gt;  GDAL does not use iconv for recoding strings.</span></span>
<span><span class="co">#&gt; GDAL binary built with GEOS: TRUE </span></span>
<span><span class="co">#&gt; Loaded PROJ runtime: Rel. 9.2.0, March 1st, 2023, [PJ_VERSION: 920]</span></span>
<span><span class="co">#&gt; Path to PROJ shared files: C:/R/R-4.3.1/library/rgdal/proj</span></span>
<span><span class="co">#&gt; PROJ CDN enabled: FALSE</span></span>
<span><span class="co">#&gt; Linking to sp version:2.0-0</span></span>
<span><span class="co">#&gt; To mute warnings of possible GDAL/OSR exportToProj4() degradation,</span></span>
<span><span class="co">#&gt; use options("rgdal_show_exportToProj4_warnings"="none") before loading sp or rgdal.</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'pegas':</span></span>
<span><span class="co">#&gt;   method      from</span></span>
<span><span class="co">#&gt;   print.amova ade4</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">#&gt;   method from   </span></span>
<span><span class="co">#&gt;   +.gg   ggplot2</span></span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,</span></span>
<span><span class="co">#&gt; which was just loaded, will retire in October 2023.</span></span>
<span><span class="co">#&gt; Please refer to R-spatial evolution reports for details, especially</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html.</span></span>
<span><span class="co">#&gt; It may be desirable to make the sf package available;</span></span>
<span><span class="co">#&gt; package maintainers should consider adding sf to Suggests:.</span></span>
<span><span class="co">#&gt; The sp package is now running under evolution status 2</span></span>
<span><span class="co">#&gt;      (status 2 uses the sf package in place of rgdal)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'genetics':</span></span>
<span><span class="co">#&gt;   method      from </span></span>
<span><span class="co">#&gt;   [.haplotype pegas</span></span>
<span><span class="co">#&gt; Please note that rgdal will be retired during October 2023,</span></span>
<span><span class="co">#&gt; plan transition to sf/stars/terra functions using GDAL and PROJ</span></span>
<span><span class="co">#&gt; at your earliest convenience.</span></span>
<span><span class="co">#&gt; See https://r-spatial.org/r/2023/05/15/evolution4.html and https://github.com/r-spatial/evolution</span></span>
<span><span class="co">#&gt; rgdal: version: 1.6-7, (SVN revision 1203)</span></span>
<span><span class="co">#&gt; Geospatial Data Abstraction Library extensions to R successfully loaded</span></span>
<span><span class="co">#&gt; Loaded GDAL runtime: GDAL 3.6.2, released 2023/01/02</span></span>
<span><span class="co">#&gt; Path to GDAL shared files: C:/R/R-4.3.1/library/rgdal/gdal</span></span>
<span><span class="co">#&gt;  GDAL does not use iconv for recoding strings.</span></span>
<span><span class="co">#&gt; GDAL binary built with GEOS: TRUE </span></span>
<span><span class="co">#&gt; Loaded PROJ runtime: Rel. 9.2.0, March 1st, 2023, [PJ_VERSION: 920]</span></span>
<span><span class="co">#&gt; Path to PROJ shared files: C:/R/R-4.3.1/library/rgdal/proj</span></span>
<span><span class="co">#&gt; PROJ CDN enabled: FALSE</span></span>
<span><span class="co">#&gt; Linking to sp version:2.0-0</span></span>
<span><span class="co">#&gt; To mute warnings of possible GDAL/OSR exportToProj4() degradation,</span></span>
<span><span class="co">#&gt; use options("rgdal_show_exportToProj4_warnings"="none") before loading sp or rgdal.</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'pegas':</span></span>
<span><span class="co">#&gt;   method      from</span></span>
<span><span class="co">#&gt;   print.amova ade4</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">#&gt;   method from   </span></span>
<span><span class="co">#&gt;   +.gg   ggplot2</span></span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,</span></span>
<span><span class="co">#&gt; which was just loaded, will retire in October 2023.</span></span>
<span><span class="co">#&gt; Please refer to R-spatial evolution reports for details, especially</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html.</span></span>
<span><span class="co">#&gt; It may be desirable to make the sf package available;</span></span>
<span><span class="co">#&gt; package maintainers should consider adding sf to Suggests:.</span></span>
<span><span class="co">#&gt; The sp package is now running under evolution status 2</span></span>
<span><span class="co">#&gt;      (status 2 uses the sf package in place of rgdal)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'genetics':</span></span>
<span><span class="co">#&gt;   method      from </span></span>
<span><span class="co">#&gt;   [.haplotype pegas</span></span>
<span><span class="co">#&gt; Please note that rgdal will be retired during October 2023,</span></span>
<span><span class="co">#&gt; plan transition to sf/stars/terra functions using GDAL and PROJ</span></span>
<span><span class="co">#&gt; at your earliest convenience.</span></span>
<span><span class="co">#&gt; See https://r-spatial.org/r/2023/05/15/evolution4.html and https://github.com/r-spatial/evolution</span></span>
<span><span class="co">#&gt; rgdal: version: 1.6-7, (SVN revision 1203)</span></span>
<span><span class="co">#&gt; Geospatial Data Abstraction Library extensions to R successfully loaded</span></span>
<span><span class="co">#&gt; Loaded GDAL runtime: GDAL 3.6.2, released 2023/01/02</span></span>
<span><span class="co">#&gt; Path to GDAL shared files: C:/R/R-4.3.1/library/rgdal/gdal</span></span>
<span><span class="co">#&gt;  GDAL does not use iconv for recoding strings.</span></span>
<span><span class="co">#&gt; GDAL binary built with GEOS: TRUE </span></span>
<span><span class="co">#&gt; Loaded PROJ runtime: Rel. 9.2.0, March 1st, 2023, [PJ_VERSION: 920]</span></span>
<span><span class="co">#&gt; Path to PROJ shared files: C:/R/R-4.3.1/library/rgdal/proj</span></span>
<span><span class="co">#&gt; PROJ CDN enabled: FALSE</span></span>
<span><span class="co">#&gt; Linking to sp version:2.0-0</span></span>
<span><span class="co">#&gt; To mute warnings of possible GDAL/OSR exportToProj4() degradation,</span></span>
<span><span class="co">#&gt; use options("rgdal_show_exportToProj4_warnings"="none") before loading sp or rgdal.</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'pegas':</span></span>
<span><span class="co">#&gt;   method      from</span></span>
<span><span class="co">#&gt;   print.amova ade4</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">#&gt;   method from   </span></span>
<span><span class="co">#&gt;   +.gg   ggplot2</span></span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,</span></span>
<span><span class="co">#&gt; which was just loaded, will retire in October 2023.</span></span>
<span><span class="co">#&gt; Please refer to R-spatial evolution reports for details, especially</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html.</span></span>
<span><span class="co">#&gt; It may be desirable to make the sf package available;</span></span>
<span><span class="co">#&gt; package maintainers should consider adding sf to Suggests:.</span></span>
<span><span class="co">#&gt; The sp package is now running under evolution status 2</span></span>
<span><span class="co">#&gt;      (status 2 uses the sf package in place of rgdal)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'genetics':</span></span>
<span><span class="co">#&gt;   method      from </span></span>
<span><span class="co">#&gt;   [.haplotype pegas</span></span>
<span><span class="co">#&gt; Please note that rgdal will be retired during October 2023,</span></span>
<span><span class="co">#&gt; plan transition to sf/stars/terra functions using GDAL and PROJ</span></span>
<span><span class="co">#&gt; at your earliest convenience.</span></span>
<span><span class="co">#&gt; See https://r-spatial.org/r/2023/05/15/evolution4.html and https://github.com/r-spatial/evolution</span></span>
<span><span class="co">#&gt; rgdal: version: 1.6-7, (SVN revision 1203)</span></span>
<span><span class="co">#&gt; Geospatial Data Abstraction Library extensions to R successfully loaded</span></span>
<span><span class="co">#&gt; Loaded GDAL runtime: GDAL 3.6.2, released 2023/01/02</span></span>
<span><span class="co">#&gt; Path to GDAL shared files: C:/R/R-4.3.1/library/rgdal/gdal</span></span>
<span><span class="co">#&gt;  GDAL does not use iconv for recoding strings.</span></span>
<span><span class="co">#&gt; GDAL binary built with GEOS: TRUE </span></span>
<span><span class="co">#&gt; Loaded PROJ runtime: Rel. 9.2.0, March 1st, 2023, [PJ_VERSION: 920]</span></span>
<span><span class="co">#&gt; Path to PROJ shared files: C:/R/R-4.3.1/library/rgdal/proj</span></span>
<span><span class="co">#&gt; PROJ CDN enabled: FALSE</span></span>
<span><span class="co">#&gt; Linking to sp version:2.0-0</span></span>
<span><span class="co">#&gt; To mute warnings of possible GDAL/OSR exportToProj4() degradation,</span></span>
<span><span class="co">#&gt; use options("rgdal_show_exportToProj4_warnings"="none") before loading sp or rgdal.</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'pegas':</span></span>
<span><span class="co">#&gt;   method      from</span></span>
<span><span class="co">#&gt;   print.amova ade4</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">#&gt;   method from   </span></span>
<span><span class="co">#&gt;   +.gg   ggplot2</span></span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,</span></span>
<span><span class="co">#&gt; which was just loaded, will retire in October 2023.</span></span>
<span><span class="co">#&gt; Please refer to R-spatial evolution reports for details, especially</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html.</span></span>
<span><span class="co">#&gt; It may be desirable to make the sf package available;</span></span>
<span><span class="co">#&gt; package maintainers should consider adding sf to Suggests:.</span></span>
<span><span class="co">#&gt; The sp package is now running under evolution status 2</span></span>
<span><span class="co">#&gt;      (status 2 uses the sf package in place of rgdal)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'genetics':</span></span>
<span><span class="co">#&gt;   method      from </span></span>
<span><span class="co">#&gt;   [.haplotype pegas</span></span>
<span><span class="co">#&gt; Please note that rgdal will be retired during October 2023,</span></span>
<span><span class="co">#&gt; plan transition to sf/stars/terra functions using GDAL and PROJ</span></span>
<span><span class="co">#&gt; at your earliest convenience.</span></span>
<span><span class="co">#&gt; See https://r-spatial.org/r/2023/05/15/evolution4.html and https://github.com/r-spatial/evolution</span></span>
<span><span class="co">#&gt; rgdal: version: 1.6-7, (SVN revision 1203)</span></span>
<span><span class="co">#&gt; Geospatial Data Abstraction Library extensions to R successfully loaded</span></span>
<span><span class="co">#&gt; Loaded GDAL runtime: GDAL 3.6.2, released 2023/01/02</span></span>
<span><span class="co">#&gt; Path to GDAL shared files: C:/R/R-4.3.1/library/rgdal/gdal</span></span>
<span><span class="co">#&gt;  GDAL does not use iconv for recoding strings.</span></span>
<span><span class="co">#&gt; GDAL binary built with GEOS: TRUE </span></span>
<span><span class="co">#&gt; Loaded PROJ runtime: Rel. 9.2.0, March 1st, 2023, [PJ_VERSION: 920]</span></span>
<span><span class="co">#&gt; Path to PROJ shared files: C:/R/R-4.3.1/library/rgdal/proj</span></span>
<span><span class="co">#&gt; PROJ CDN enabled: FALSE</span></span>
<span><span class="co">#&gt; Linking to sp version:2.0-0</span></span>
<span><span class="co">#&gt; To mute warnings of possible GDAL/OSR exportToProj4() degradation,</span></span>
<span><span class="co">#&gt; use options("rgdal_show_exportToProj4_warnings"="none") before loading sp or rgdal.</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'pegas':</span></span>
<span><span class="co">#&gt;   method      from</span></span>
<span><span class="co">#&gt;   print.amova ade4</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">#&gt;   method from   </span></span>
<span><span class="co">#&gt;   +.gg   ggplot2</span></span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,</span></span>
<span><span class="co">#&gt; which was just loaded, will retire in October 2023.</span></span>
<span><span class="co">#&gt; Please refer to R-spatial evolution reports for details, especially</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html.</span></span>
<span><span class="co">#&gt; It may be desirable to make the sf package available;</span></span>
<span><span class="co">#&gt; package maintainers should consider adding sf to Suggests:.</span></span>
<span><span class="co">#&gt; The sp package is now running under evolution status 2</span></span>
<span><span class="co">#&gt;      (status 2 uses the sf package in place of rgdal)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'genetics':</span></span>
<span><span class="co">#&gt;   method      from </span></span>
<span><span class="co">#&gt;   [.haplotype pegas</span></span>
<span><span class="co">#&gt; Please note that rgdal will be retired during October 2023,</span></span>
<span><span class="co">#&gt; plan transition to sf/stars/terra functions using GDAL and PROJ</span></span>
<span><span class="co">#&gt; at your earliest convenience.</span></span>
<span><span class="co">#&gt; See https://r-spatial.org/r/2023/05/15/evolution4.html and https://github.com/r-spatial/evolution</span></span>
<span><span class="co">#&gt; rgdal: version: 1.6-7, (SVN revision 1203)</span></span>
<span><span class="co">#&gt; Geospatial Data Abstraction Library extensions to R successfully loaded</span></span>
<span><span class="co">#&gt; Loaded GDAL runtime: GDAL 3.6.2, released 2023/01/02</span></span>
<span><span class="co">#&gt; Path to GDAL shared files: C:/R/R-4.3.1/library/rgdal/gdal</span></span>
<span><span class="co">#&gt;  GDAL does not use iconv for recoding strings.</span></span>
<span><span class="co">#&gt; GDAL binary built with GEOS: TRUE </span></span>
<span><span class="co">#&gt; Loaded PROJ runtime: Rel. 9.2.0, March 1st, 2023, [PJ_VERSION: 920]</span></span>
<span><span class="co">#&gt; Path to PROJ shared files: C:/R/R-4.3.1/library/rgdal/proj</span></span>
<span><span class="co">#&gt; PROJ CDN enabled: FALSE</span></span>
<span><span class="co">#&gt; Linking to sp version:2.0-0</span></span>
<span><span class="co">#&gt; To mute warnings of possible GDAL/OSR exportToProj4() degradation,</span></span>
<span><span class="co">#&gt; use options("rgdal_show_exportToProj4_warnings"="none") before loading sp or rgdal.</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'pegas':</span></span>
<span><span class="co">#&gt;   method      from</span></span>
<span><span class="co">#&gt;   print.amova ade4</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">#&gt;   method from   </span></span>
<span><span class="co">#&gt;   +.gg   ggplot2</span></span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,</span></span>
<span><span class="co">#&gt; which was just loaded, will retire in October 2023.</span></span>
<span><span class="co">#&gt; Please refer to R-spatial evolution reports for details, especially</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html.</span></span>
<span><span class="co">#&gt; It may be desirable to make the sf package available;</span></span>
<span><span class="co">#&gt; package maintainers should consider adding sf to Suggests:.</span></span>
<span><span class="co">#&gt; The sp package is now running under evolution status 2</span></span>
<span><span class="co">#&gt;      (status 2 uses the sf package in place of rgdal)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'genetics':</span></span>
<span><span class="co">#&gt;   method      from </span></span>
<span><span class="co">#&gt;   [.haplotype pegas</span></span>
<span><span class="co">#&gt; Please note that rgdal will be retired during October 2023,</span></span>
<span><span class="co">#&gt; plan transition to sf/stars/terra functions using GDAL and PROJ</span></span>
<span><span class="co">#&gt; at your earliest convenience.</span></span>
<span><span class="co">#&gt; See https://r-spatial.org/r/2023/05/15/evolution4.html and https://github.com/r-spatial/evolution</span></span>
<span><span class="co">#&gt; rgdal: version: 1.6-7, (SVN revision 1203)</span></span>
<span><span class="co">#&gt; Geospatial Data Abstraction Library extensions to R successfully loaded</span></span>
<span><span class="co">#&gt; Loaded GDAL runtime: GDAL 3.6.2, released 2023/01/02</span></span>
<span><span class="co">#&gt; Path to GDAL shared files: C:/R/R-4.3.1/library/rgdal/gdal</span></span>
<span><span class="co">#&gt;  GDAL does not use iconv for recoding strings.</span></span>
<span><span class="co">#&gt; GDAL binary built with GEOS: TRUE </span></span>
<span><span class="co">#&gt; Loaded PROJ runtime: Rel. 9.2.0, March 1st, 2023, [PJ_VERSION: 920]</span></span>
<span><span class="co">#&gt; Path to PROJ shared files: C:/R/R-4.3.1/library/rgdal/proj</span></span>
<span><span class="co">#&gt; PROJ CDN enabled: FALSE</span></span>
<span><span class="co">#&gt; Linking to sp version:2.0-0</span></span>
<span><span class="co">#&gt; To mute warnings of possible GDAL/OSR exportToProj4() degradation,</span></span>
<span><span class="co">#&gt; use options("rgdal_show_exportToProj4_warnings"="none") before loading sp or rgdal.</span></span>
<span></span>
<span><span class="va">end_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">end_time</span> <span class="op">-</span> <span class="va">start_time</span></span>
<span><span class="co">#&gt; Time difference of 3.050072 mins</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Total Duration:"</span>, <span class="va">end_time</span> <span class="op">-</span> <span class="va">start_time</span><span class="op">)</span></span>
<span><span class="co">#&gt; Total Duration: 3.050072</span></span></code></pre></div>
<p>Firstly, negative values of Fst are incorrect, and usually reflect
numerical inaccuracy around zero, so we will set negative Fst values to
zero. Then let’s look at the distribution.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_fsts</span><span class="op">[</span><span class="va">all_fsts</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0.0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">all_fsts</span>, breaks <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p><img src="simple_example_using_migration_and_fst_files/figure-html/fst_dist-1.png" width="700">
So how is Fst related to migration rate and population size? We can get
an idea by running a simple linear model. Then we will do plots.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">errored</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">param_df</span> <span class="op">&lt;-</span> <span class="va">param_df</span><span class="op">[</span><span class="op">-</span><span class="va">errored</span>, <span class="op">]</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">fst_dat</span> <span class="op">&lt;-</span> <span class="va">param_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>fst <span class="op">=</span> <span class="va">all_fsts</span>,</span>
<span>                pop_size_st <span class="op">=</span> <span class="op">(</span><span class="va">pop_size</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">pop_size</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">pop_size</span><span class="op">)</span>,</span>
<span>                m_st <span class="op">=</span> <span class="op">(</span><span class="va">m</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="va">fst</span> <span class="op">~</span> <span class="va">pop_size_st</span><span class="op">*</span><span class="va">m_st</span>, data <span class="op">=</span> <span class="va">fst_dat</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mod</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call:</span></span>
<span><span class="co">#&gt; lm(formula = fst ~ pop_size_st * m_st, data = fst_dat)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Residuals:</span></span>
<span><span class="co">#&gt;      Min       1Q   Median       3Q      Max </span></span>
<span><span class="co">#&gt; -0.09117 -0.03810 -0.00932  0.01505  0.70214 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Coefficients:</span></span>
<span><span class="co">#&gt;                   Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span><span class="co">#&gt; (Intercept)       0.029191   0.004567   6.391 6.38e-10 ***</span></span>
<span><span class="co">#&gt; pop_size_st      -0.005577   0.004597  -1.213    0.226    </span></span>
<span><span class="co">#&gt; m_st             -0.036193   0.004599  -7.870 6.72e-14 ***</span></span>
<span><span class="co">#&gt; pop_size_st:m_st  0.005673   0.004513   1.257    0.210    </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Residual standard error: 0.07911 on 296 degrees of freedom</span></span>
<span><span class="co">#&gt; Multiple R-squared:  0.1775, Adjusted R-squared:  0.1692 </span></span>
<span><span class="co">#&gt; F-statistic:  21.3 on 3 and 296 DF,  p-value: 1.616e-12</span></span></code></pre></div>
<p>So both migration rate</p>
<p>We can demonstrate the interaction by splitting our data into
population size bins and plotting the migration rate relationship as
separate lines for each bin.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fst_dat</span> <span class="op">&lt;-</span> <span class="va">fst_dat</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>ps_bin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">pop_size</span>, <span class="fl">5</span>, ordered_result <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">fst_dat</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">m</span>, <span class="va">fst</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">ps_bin</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">ps_bin</span>, fill <span class="op">=</span> <span class="va">ps_bin</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_sqrt</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre class="fansi fansi-message"><code><span><span class="co">#&gt; `geom_smooth()` using method = 'loess' and formula = 'y ~ x'</span></span></code></pre>
<p><img src="simple_example_using_migration_and_fst_files/figure-html/plot_that-1.png" width="700"></p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Russell Dinnage, Stephen Sarre, Bernd Gruber, Richard Duncan.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
