<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>SLiM method initializeSLiMOptions — initializeSLiMOptions • slimr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="SLiM method initializeSLiMOptions — initializeSLiMOptions"><meta property="og:description" content="Documentation for SLiM function initializeSLiMOptions, which is a method
of the SLiM class Initialize.
Note that the R function is a stub, it does not do anything in R (except bring
up this documentation). It will only do
anything useful when used inside a slim_block function further
nested in a slim_script
function call, where it will be translated into valid SLiM code as part of a
full SLiM script."><meta property="og:image" content="https://rdinnager.github.io/slimr/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">slimr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Custom_vis.html">Custom Live Plotting for SLiM</a>
    </li>
    <li>
      <a href="../articles/simple_example_using_migration_and_fst.html">simple_example_using_migration_and_fst</a>
    </li>
    <li>
      <a href="../articles/installing_SLiM_for_slimr.html">How to Install SLiM for Use with slimr</a>
    </li>
    <li>
      <a href="../articles/simple_nucleotide_example.html">Simple Example of a Simulation using Nucleotides</a>
    </li>
    <li>
      <a href="../articles/slimrlang_intro.html">Introduction to slimr: Writing SLiM scripts from R</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/rdinnager/slimr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>SLiM method initializeSLiMOptions</h1>
    <small class="dont-index">Source: <a href="https://github.com/rdinnager/slimr/blob/HEAD/R/slim_lang.R" class="external-link"><code>R/slim_lang.R</code></a></small>
    <div class="hidden name"><code>initializeSLiMOptions.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Documentation for SLiM function <code>initializeSLiMOptions</code>, which is a method
of the SLiM class <code><a href="Initialize.html">Initialize</a></code>.
Note that the R function is a stub, it does not do anything in R (except bring
up this documentation). It will only do
anything useful when used inside a <code><a href="slim_block.html">slim_block</a></code> function further
nested in a <code><a href="slim_script.html">slim_script</a></code>
function call, where it will be translated into valid SLiM code as part of a
full SLiM script.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">initializeSLiMOptions</span><span class="op">(</span></span>
<span>  <span class="va">keepPedigrees</span>,</span>
<span>  <span class="va">dimensionality</span>,</span>
<span>  <span class="va">periodicity</span>,</span>
<span>  <span class="va">mutationRuns</span>,</span>
<span>  <span class="va">preventIncidentalSelfing</span>,</span>
<span>  <span class="va">nucleotideBased</span>,</span>
<span>  <span class="va">randomizeCallbacks</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>keepPedigrees</dt>
<dd><p>An object of type logical or string or string or integer
or logical or logical or logical. Must be of length 1 (a singleton). The default
value is <code>F</code>. See details for description.</p></dd>


<dt>dimensionality</dt>
<dd><p>An object of type logical or string or string or integer
or logical or logical or logical. Must be of length 1 (a singleton). The default
value is <code>""</code>. See details for description.</p></dd>


<dt>periodicity</dt>
<dd><p>An object of type logical or string or string or integer or
logical or logical or logical. Must be of length 1 (a singleton). The default
value is <code>""</code>. See details for description.</p></dd>


<dt>mutationRuns</dt>
<dd><p>An object of type logical or string or string or integer or
logical or logical or logical. Must be of length 1 (a singleton). The default
value is <code>0</code>. See details for description.</p></dd>


<dt>preventIncidentalSelfing</dt>
<dd><p>An object of type logical or string or string or
integer or logical or logical or logical. Must be of length 1 (a singleton). The
default value is <code>F</code>. See details for description.</p></dd>


<dt>nucleotideBased</dt>
<dd><p>An object of type logical or string or string or integer
or logical or logical or logical. Must be of length 1 (a singleton). The default
value is <code>F</code>. See details for description.</p></dd>


<dt>randomizeCallbacks</dt>
<dd><p>An object of type logical or string or string or
integer or logical or logical or logical. Must be of length 1 (a singleton). The
default value is <code>F</code>. See details for description.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An object of type void.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Documentation for this function can be found in the official
<a href="http://benhaller.com/slim/SLiM_Manual.pdf#page=625" class="external-link">SLiM manual: page
625</a>.</p>
<p>Configure options for the simulation. If initializeSLiMOptions()
is called at all then it must be called before any other initialization
function (except initializeSLiMModelType()), so that SLiM knows from the
outset which optional features are enabled and which are not. If keepPedigrees
is T, SLiM will keep pedigree information for every individual in the
simulation, tracking the identity of its parents and grandparents. This
allows individuals to assess their degree of pedigree-based relatedness to
other individuals (see Individual's relatedness() method, section 24.7.2), as
well as allowing a model to find "trios" (two parents and an offspring they
generated) using the pedigree properties of Individual (section 24.7.1). As a
side effect of keepPedigrees being T, the pedigreeID, pedigreeParentIDs, and
pedigreeGrandparentIDs properties of Individual will have defined values (see
section 24.7.1), as will the genomePedigreeID property of Genome (see section
24.4.1). Note that pedigree-based relatedness doesn't necessarily correspond
to genetic relatedness, due to effects such as assortment and recombination.
For an overview of other ways of tracking genetic ancestry, including true
local ancestry at each position on the chromosome, see sections 1.7 and 14.7.
Beginning in SLiM 3.5, keepPedigrees=T also enables tracking of individual
reproductive output, available through the reproductiveOutput property of
Individual (see section 24.7.1) and the lifetimeReproductiveOutput property
of Subpopulation (see section 24.15.1). If dimensionality is not "", SLiM will
enable its optional "continuous space" facility. Three values for dimensionality
are presently supported: "x", "xy", and "xyz", specifying that continuous
space should be enabled for one, two, or three dimensions, respectively, using
(x), (x, y), and (x, y, z) coordinates respectively. This has a number of side
effects. First of all, it means that the specified properties of Individual (x,
y, and/or z) will be interpreted by SLiM as spatial positions; in particular,
SLiMgui will use those properties to display subpopulations spatially. Second,
it allows spatial interactions to be defined, evaluated, and queried using
initializeInteractionType() and interaction() callbacks. And third, it enables
the use of any other properties and methods related to continuous space, such
as setting the spatial boundaries of subpopulations, which would otherwise
raise an error. If periodicity is not "", SLiM will designate the specified
spatial dimensions as being periodic - wrapping around at the edges of the
spatial boundaries of that dimension. This option may only be used if the
dimensionality parameter to initializeSLiMOptions() has been used to enable
spatiality in the model, and only spatial dimensions that were specified in the
dimensionality of the model may be declared to be periodic (but if desired, it
is permissible to make just a subset of those dimensions periodic; it is not
an all-or-none proposition). For example, if the specified dimensionality is
"xy", the model's periodicity may be "x", "y", or "xy" (or "", the default,
to specify that there are no periodic dimensions). A one-dimensional periodic
model would model a space like the perimeter of a circle. A two-dimensional
model periodic in one of those dimensions would model a space like a cylinder
without its end caps; if periodic in both dimensions, the modeled space is a
torus. The shapes of three-dimensional periodic models are harder to visualize,
but are essentially higherdimensional analogues of these concepts. Periodic
boundary conditions are commonly used to model spatial scenarios without "edge
effects", since there are no edges in the periodic spatial dimensions. The
pointPeriodic() method of Subpopulation is typically used in conjunction with
this option, to actually implement the periodic boundary condition for the
specified dimensions. If mutationRuns is not 0, SLiM will use the value given
as the number of mutation runs inside Genome objects; if it is 0 (the default),
SLiM will calculate a number of mutation runs that it estimates will work
well. Internally, SLiM divides genomes into a sequence of consecutive mutation
runs, allowing more efficient internal computations. The optimal mutation
run length is short enough that each mutation run is relatively unlikely to
be modified by mutation/recombination events when inherited, but long enough
that each mutation run is likely to contain a relatively large number of
mutations; these priorities are in tension, so an intermediate balance between
them is generally desirable. The optimal number of mutation runs will depend
upon the machine and even the compiler used to build SLiM, so SLiM's default
value may not be optimal; for maximal performance it can thus be beneficial to
experiment with different values and find the optimal value for the simulation
- a process which SLiM can assist with (see section 21.4). Specifying the
number of mutation runs is an advanced technique, but in certain cases it can
improve performance significantly. If preventIncidentalSelfing is T, incidental
selfing in hermaphroditic models will be prevented by SLiM. By default (i.e.,
if preventIncidentalSelfing is F), SLiM chooses the first and second parents
in a biparental mating event independently. It is therefore possible for the
same individual to be chosen as both the first and second parent, resulting
in selfing events even when the selfing rate is zero. In many models this
is unimportant, since it happens fairly infrequently and does not have large
consequences. This behavior is SLiM's default because it is the simplest option,
and produces results that most closely align with simple analytical population
genetics models. However, in some models this selfing can be undesirable and
problematic. In particular, models that involve very high variance in fitness
or very small effective population sizes may see elevated rates of selfing that
substantially influence model results. If preventIncidentalSelfing is set to T,
all such incidental selfing will be prevented (by choosing a new second parent
if the first parent was chosen again). Non-incidental selfing, as requested by
the selfing rate, will still be permitted. Note that if incidental selfing is
prevented, SLiM will hang if it is unable to find a different second parent;
there must always be at least two individuals in the population with non-zero
fitness, and mateChoice() and modifyChild() callbacks must not absolutely
prevent those two individuals from producing viable offspring. Enforcement
of the prohibition on incidental selfing will occur after mateChoice()
callbacks have been called (and thus the default mating weights provided to
mateChoice() callbacks will not exclude the first parent!), but will occur
before modifyChild() callbacks are called (so those callbacks may assume that
the first and second parents are distinct). If nucleotideBased is T, the model
will be nucleotide-based. In this case, auto-generated mutations (i.e., mutation
types used by genomic element types) must be nucleotide-based, and an ancestral
nucleotide sequence must be supplied with initializeAncestralNucleotides().
Nonnucleotide- based mutations may still be used, but may not be referenced by
genomic element types. A mutation rate (or rate map) may not be supplied with
initializeMutationRate(); instead, a hotspot map may (optionally) be supplied
with initializeHotspotMap(). This choice has many consequences across SLiM; see
section 1.8 for further discussion. If randomizeCallbacks is T (the default),
the order in which individuals are processed in callbacks will be randomized
to make it easier to avoid order-dependency bugs. This flag exists because the
order of individuals in each subpopulation is non-random; most notably, females
always come before males in the individuals vector, but non-random ordering may
also occur with respect to things like migrant versus non-migrant status, origin
by selfing versus cloning versus biparental mating, and other factors. When this
option is F, individuals in a subpopulation are processed in the order of the
individuals vector in each tick cycle stage, which may lead to order-dependency
issues if there is an enabled callback whose behavior is not fully independent
between calls. Setting this option to T will cause individuals within each
subpopulation to be processed in a randomized order in each tick cycle stage;
specifically, this randomizes the order of calls to mutationEffect() callbacks
in both WF and nonWF models, and calls to reproduction() and survival()
callbacks in nonWF models. Each subpopulation is still processed separately,
in sequential order, so order-dependency issues between subpopulations are
still possible if callbacks have effects that are not fully independent. This
feature was added in SLiM 4, breaking backward compatibility; to recover the
behavior of previous versions of SLiM, pass F for this option (but then be very
careful about order-dependency issues in your script). The default of T is the
safe option, but a small speed penalty is incurred by the randomization of the
processing order - for most models the difference will be less than 1<!-- %, but in -->
the worst case it may approach 10<!-- %. Models that do not have any order-dependency -->
issue may therefore run somewhat faster if this is set to F. Note that anywhere
that your script uses the individuals property of Subpopulation, the order
of individuals returned will be non-random (regardless of the setting of this
option); you should use sample() to shuffle the order of the individuals vector
if necessary to avoid order-dependency issues in your script. This function
will likely be extended with further options in the future, added on to the end
of the argument list. Using named arguments with this call is recommended for
readability. Note that turning on optional features may increase the runtime and
memory footprint of SLiM.</p>
    </div>
    <div id="copyright">
    <h2>Copyright</h2>
    

<p>This is documentation for a function in the SLiM software, and has been
reproduced from the official manual,
which can be found here: <a href="http://benhaller.com/slim/SLiM_Manual.pdf" class="external-link">http://benhaller.com/slim/SLiM_Manual.pdf</a>. This
documentation is
Copyright © 2016-2020 Philipp Messer. All rights reserved. More information
about SLiM can be found
on the official website: <a href="https://messerlab.org/slim/" class="external-link">https://messerlab.org/slim/</a></p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Other Initialize: 
<code><a href="Initialize.html">Init</a></code>,
<code><a href="initializeAncestralNucleotides.html">initializeAncestralNucleotides</a>()</code>,
<code><a href="initializeGeneConversion.html">initializeGeneConversion</a>()</code>,
<code><a href="initializeGenomicElementType.html">initializeGenomicElementType</a>()</code>,
<code><a href="initializeGenomicElement.html">initializeGenomicElement</a>()</code>,
<code><a href="initializeHotspotMap.html">initializeHotspotMap</a>()</code>,
<code><a href="initializeInteractionType.html">initializeInteractionType</a>()</code>,
<code><a href="initializeMutationRate.html">initializeMutationRate</a>()</code>,
<code><a href="initializeMutationTypeNuc.html">initializeMutationTypeNuc</a>()</code>,
<code><a href="initializeMutationType.html">initializeMutationType</a>()</code>,
<code><a href="initializeRecombinationRate.html">initializeRecombinationRate</a>()</code>,
<code><a href="initializeSLiMModelType.html">initializeSLiMModelType</a>()</code>,
<code><a href="initializeSex.html">initializeSex</a>()</code>,
<code><a href="initializeSpecies.html">initializeSpecies</a>()</code>,
<code><a href="initializeTreeSeq.html">initializeTreeSeq</a>()</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Benjamin C Haller (<a href="mailto:bhaller@benhaller.com">bhaller@benhaller.com</a>) and Philipp W Messer
(<a href="mailto:messer@cornell.edu">messer@cornell.edu</a>)</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Russell Dinnage, Stephen Sarre, Bernd Gruber, Richard Duncan.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

