<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>SLiM method nucleotidesToCodons — nucleotidesToCodons • slimr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="SLiM method nucleotidesToCodons — nucleotidesToCodons"><meta property="og:description" content="Documentation for SLiM function nucleotidesToCodons, which is a method of
the SLiM class SLiMBuiltin.
Note that the R function is a stub, it does not do anything in R (except bring
up this documentation). It will only do
anything useful when used inside a slim_block function further
nested in a slim_script
function call, where it will be translated into valid SLiM code as part of a
full SLiM script."><meta property="og:image" content="https://rdinnager.github.io/slimr/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">slimr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Custom_vis.html">Custom Live Plotting for SLiM</a>
    </li>
    <li>
      <a href="../articles/simple_example_using_migration_and_fst.html">simple_example_using_migration_and_fst</a>
    </li>
    <li>
      <a href="../articles/installing_SLiM_for_slimr.html">How to Install SLiM for Use with slimr</a>
    </li>
    <li>
      <a href="../articles/simple_nucleotide_example.html">Simple Example of a Simulation using Nucleotides</a>
    </li>
    <li>
      <a href="../articles/slimrlang_intro.html">Introduction to slimr: Writing SLiM scripts from R</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/rdinnager/slimr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>SLiM method nucleotidesToCodons</h1>
    <small class="dont-index">Source: <a href="https://github.com/rdinnager/slimr/blob/HEAD/R/slim_lang.R" class="external-link"><code>R/slim_lang.R</code></a></small>
    <div class="hidden name"><code>nucleotidesToCodons.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Documentation for SLiM function <code>nucleotidesToCodons</code>, which is a method of
the SLiM class <code><a href="SLiMBuiltin.html">SLiMBuiltin</a></code>.
Note that the R function is a stub, it does not do anything in R (except bring
up this documentation). It will only do
anything useful when used inside a <code><a href="slim_block.html">slim_block</a></code> function further
nested in a <code><a href="slim_script.html">slim_script</a></code>
function call, where it will be translated into valid SLiM code as part of a
full SLiM script.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">nucleotidesToCodons</span><span class="op">(</span><span class="va">sequence</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>sequence</dt>
<dd><p>An object of type integer or string. See details for
description.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An object of type integer.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Documentation for this function can be found in the official <a href="http://%0Abenhaller.com/slim/SLiM_Manual.pdf#page=591" class="external-link">SLiM manual: page 591</a>.</p>
<p>Returns the codon sequence corresponding to the nucleotide sequence
in sequence. The codon sequence is an integer vector with values from 0 to
63, based upon successive nucleotide triplets in the nucleotide sequence. The
codon value for a given nucleotide triplet XYZ is 16X + 4Y + Z, where X, Y, and
Z have the usual values A=0, C=1, G=2, T=3. For example, the triplet AAA has
a codon value of 0, AAC is 1, AAG is 2, AAT is 3, ACA is 4, and on upward to
TTT which is 63. If the nucleotide sequence AACACATTT is passed in, the codon
vector 1 4 63 will therefore be returned. These codon values can be useful
in themselves; they can also be passed to codonsToAminoAcids() to translate
them into the corresponding amino acid sequence if desired. The nucleotide
sequence in sequence may be supplied in any of three formats: a string vector
with single-letter nucleotides (e.g., "T", "A", "T", "A"), a singleton string
of nucleotide letters (e.g., "TATA"), or an integer vector of nucleotide
values (e.g., 3, 0, 3, 0) using SLiM's standard code of A=0, C=1, G=2, T=3.
If the choice of format is not driven by other considerations, such as ease
of manipulation, then the singleton string format will certainly be the most
memory-efficient for long sequences, and will probably also be the fastest.
The nucleotide sequence provided must be a multiple of three in length, so that
it translates to an integral number of codons. 0 α α α α 0 α α α α 0 α α α α
0 0 β α β β 0 β α α β 0 β β α β 0 (is)randomNucleotides(integer$ length, [Nif
basis = NULL], [string$ format = "string"]) Generates a new random nucleotide
sequence with length bases. The four nucleotides ACGT are equally probable
if basis is NULL (the default); otherwise, basis may be a 4-element integer
or float vector providing relative fractions for A, C, G, and T respectively
(these need not sum to 1.0, as they will be normalized). More complex generative
models such as Markov processes are not supported intrinsically in SLiM at
this time, but arbitrary generated sequences may always be loaded from files
on disk. The format parameter controls the format of the returned sequence. It
may be "string" to obtain the generated sequence as a singleton string (e.g.,
"TATA"), "char" to obtain it as a string vector of single characters (e.g.,
"T", "A", "T", "A"), or "integer" to obtain it as an integer vector (e.g., 3,
0, 3, 0), using SLiM's standard code of A=0, C=1, G=2, T=3. For passing directly
to initializeAncestralNucleotides(), format "string" (a singleton string) will
certainly be the most memory-efficient, and probably also the fastest. Memory
efficiency can be a significant consideration; the nucleotide sequence for a
chromosome of length 109 will occupy approximately 1 GB of memory when stored
as a singleton string (with one byte per nucleotide), and much more if stored
in the other formats. However, the other formats can be easier to work with
in Eidos, and so may be preferable for relatively short chromosomes if you are
manipulating the generated sequence.</p>
    </div>
    <div id="copyright">
    <h2>Copyright</h2>
    

<p>This is documentation for a function in the SLiM software, and has been
reproduced from the official manual,
which can be found here: <a href="http://benhaller.com/slim/SLiM_Manual.pdf" class="external-link">http://benhaller.com/slim/SLiM_Manual.pdf</a>. This
documentation is
Copyright © 2016-2020 Philipp Messer. All rights reserved. More information
about SLiM can be found
on the official website: <a href="https://messerlab.org/slim/" class="external-link">https://messerlab.org/slim/</a></p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Other SLiMBuiltin: 
<code><a href="SLiMBuiltin.html">SLiMBuiltin</a></code>,
<code><a href="calcFST.html">calcFST</a>()</code>,
<code><a href="calcHeterozygosity.html">calcHeterozygosity</a>()</code>,
<code><a href="calcPairHeterozygosity.html">calcPairHeterozygosity</a>()</code>,
<code><a href="calcVA.html">calcVA</a>()</code>,
<code><a href="calcWattersonsTheta.html">calcWattersonsTheta</a>()</code>,
<code><a href="codonsToAminoAcids.html">codonsToAminoAcids</a>()</code>,
<code><a href="mm16To256.html">mm16To256</a>()</code>,
<code><a href="mmJukesCantor.html">mmJukesCantor</a>()</code>,
<code><a href="mmKimura.html">mmKimura</a>()</code>,
<code><a href="nucleotideCounts.html">nucleotideCounts</a>()</code>,
<code><a href="nucleotideFrequencies.html">nucleotideFrequencies</a>()</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Benjamin C Haller (<a href="mailto:bhaller@benhaller.com">bhaller@benhaller.com</a>) and Philipp W Messer
(<a href="mailto:messer@cornell.edu">messer@cornell.edu</a>)</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Russell Dinnage, Stephen Sarre, Bernd Gruber, Richard Duncan.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

