<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Create a file to initialise a population in SLiM — slim_make_pop_input • slimr</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Create a file to initialise a population in SLiM — slim_make_pop_input" />
<meta property="og:description" content="Make a SLiM file in the style of outputFull() which can be read by SLiM. This is useful to
start a simulation with a particular population state that you define in R" />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">slimr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9001</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/articles/Custom_vis.html">Custom Live Plotting for SLiM</a>
    </li>
    <li>
      <a href="../articles/articles/simple_example_using_migration_and_fst.html">simple_example_using_migration_and_fst</a>
    </li>
    <li>
      <a href="../articles/installing_SLiM_for_slimr.html">How to Install SLiM for Use with slimr</a>
    </li>
    <li>
      <a href="../articles/slimrlang_intro.html">Introduction to slimr: Writing SLiM scripts from R</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rdinnager/slimr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create a file to initialise a population in SLiM</h1>
    <small class="dont-index">Source: <a href='https://github.com/rdinnager/slimr/blob/master/R/slim_input.R'><code>R/slim_input.R</code></a></small>
    <div class="hidden name"><code>slim_make_pop_input.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Make a SLiM file in the style of <code><a href='outputFull.html'>outputFull()</a></code> which can be read by SLiM. This is useful to
start a simulation with a particular population state that you define in R</p>
    </div>

    <pre class="usage"><span class='fu'>slim_make_pop_input</span><span class='op'>(</span>
  <span class='va'>snps</span>,
  file_name <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempfile</a></span><span class='op'>(</span><span class='op'>)</span>,
  sim_gen <span class='op'>=</span> <span class='fl'>10000</span>,
  ind_pops <span class='op'>=</span> <span class='cn'>NULL</span>,
  ind_sex <span class='op'>=</span> <span class='cn'>NULL</span>,
  mut_pos <span class='op'>=</span> <span class='cn'>NULL</span>,
  mut_type <span class='op'>=</span> <span class='cn'>NULL</span>,
  mut_sel <span class='op'>=</span> <span class='cn'>NULL</span>,
  mut_dom <span class='op'>=</span> <span class='cn'>NULL</span>,
  mut_pop <span class='op'>=</span> <span class='cn'>NULL</span>,
  mut_gen <span class='op'>=</span> <span class='cn'>NULL</span>,
  mut_nuc <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>snps</th>
      <td><p>R object containing SNPs. Can be a matrix with 0, 1, or 2 as its elements,
or a <code>genlight</code> object. Rows correspond to individuals, columns to loci.</p></td>
    </tr>
    <tr>
      <th>file_name</th>
      <td><p>Path to file where the population input file should be saved.</p></td>
    </tr>
    <tr>
      <th>sim_gen</th>
      <td><p>What generation to write in the file. This doesn't really do anything, but is just a requirement
for the format.</p></td>
    </tr>
    <tr>
      <th>ind_pops</th>
      <td><p>An optional character vector with length equal to <code><a href='https://rdrr.io/r/base/nrow.html'>nrow(snps)</a></code> with Subpopulation indicators for
each individual. Should be "p1", "p2", etc. as this is how subpopulations are named in SLiM.</p></td>
    </tr>
    <tr>
      <th>ind_sex</th>
      <td><p>An optional character vector with length equal to <code><a href='https://rdrr.io/r/base/nrow.html'>nrow(snps)</a></code> with the sex for each individual (can be
"H" for hermaphrodite, "F" for female, or "M" for male.)</p></td>
    </tr>
    <tr>
      <th>mut_pos</th>
      <td><p>An optional integer vector with length equal to <code><a href='https://rdrr.io/r/base/nrow.html'>ncol(snps)</a></code> specifying the positions in the genome
of each loci.</p></td>
    </tr>
    <tr>
      <th>mut_type</th>
      <td><p>An optional character vector with length equal to <code><a href='https://rdrr.io/r/base/nrow.html'>nrow(snps)</a></code> with the mutation type of each loci.
Should be "m1", "m2", etc. as this is how mutation types are named in SLiM.</p></td>
    </tr>
    <tr>
      <th>mut_sel</th>
      <td><p>An optional numeric vector with length equal to <code><a href='https://rdrr.io/r/base/nrow.html'>nrow(snps)</a></code> with the selection coefficient for the
mutation type of each loci.</p></td>
    </tr>
    <tr>
      <th>mut_dom</th>
      <td><p>An optional numeric vector with length equal to <code><a href='https://rdrr.io/r/base/nrow.html'>nrow(snps)</a></code> with the dominance coefficient for the
mutation type of each loci.</p></td>
    </tr>
    <tr>
      <th>mut_pop</th>
      <td><p>An optional character vector with length equal to <code><a href='https://rdrr.io/r/base/nrow.html'>nrow(snps)</a></code> with the subpopulation of origin
for the mutation at each loci. Should be "p1", "p2", etc. as this is how subpopulations are named in SLiM.</p></td>
    </tr>
    <tr>
      <th>mut_gen</th>
      <td><p>An optional integer vector with length equal to <code><a href='https://rdrr.io/r/base/nrow.html'>ncol(snps)</a></code> specifying the generation of origin
for the mutation at each loci.</p></td>
    </tr>
    <tr>
      <th>mut_nuc</th>
      <td><p>An optional character vector with length equal to <code><a href='https://rdrr.io/r/base/nrow.html'>nrow(snps)</a></code> with the nucleotide of the
mutation. Should be "A", "C", "G", or "T".</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    



  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Russell Dinnage, Stephen Sarre, Bernd Gruber, Richard Duncan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


